{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PZQ-R field variants\n",
    "\n",
    "\n",
    "We score and investigate variants from natural populations in order to have a comprehensive view of those that could impact praziquantel efficacy. We used data from different locations in New and Old World:\n",
    "* Brazil (original from the associated publication)\n",
    "* Senegal\n",
    "* Niger\n",
    "* Tanzania\n",
    "* Uganda (data generated by Sanger https://doi.org/10.3389/fgene.2019.00826)\n",
    "* Oman\n",
    "\n",
    "\n",
    "## Environment and data\n",
    "\n",
    "### Environment\n",
    "\n",
    "Creating a conda environment improves reproducibility by installing specific versions of the programs used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-01-06T16:09:29.118Z"
    }
   },
   "outputs": [],
   "source": [
    "conda env create -f .env/env.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cell below must be run each time a new Jupyter session is run or when the kernel is rebooted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-22T02:13:24.853986Z",
     "start_time": "2021-01-22T02:13:15.454757Z"
    }
   },
   "outputs": [],
   "source": [
    "# Activate the environment\n",
    "source $(sed \"s,/bin/conda,,\" <<<$CONDA_EXE)/etc/profile.d/conda.sh\n",
    "conda activate PZQ-R_field\n",
    "\n",
    "# Remove potential variable interferences\n",
    "export PERL5LIB=\"\"\n",
    "export PYTHONNOUSERSITE=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cell below must be run only once at the time of the environment creation. The R script install the R dependencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T22:49:52.915105Z",
     "start_time": "2021-01-21T22:49:13.794595Z"
    }
   },
   "outputs": [],
   "source": [
    "# Installing needed R packages\n",
    "Rscript \".env/R package dependencies.R\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Phred/Phrap/Consed and PolyPhred\n",
    "\n",
    "[Phred/Phrap/Consed](http://www.phrap.org/phredphrapconsed.html) and [PolyPhred](https://droog.gs.washington.edu/polyphred/) are software used to analyze Sanger sequencing traces by performing alignment and genotyping, respectively. They are freely available for academic used. However, this requires to contact the authors in order to get a copy ([here](http://www.phrap.org/consed/consed.html#howToGet) for Phred/Phrap/Consed, [here](https://droog.gs.washington.edu/polyphred/poly_get.html) for PolyPhred).\n",
    "\n",
    "Once the copies obtained, they must be place in a download folder at the same level as this notebook. Then programs can be compiled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-09T23:49:34.438439Z",
     "start_time": "2021-02-09T23:49:34.224809Z"
    }
   },
   "outputs": [],
   "source": [
    "# Folder in which the software archives must be located\n",
    "ddir=\"download\"\n",
    "[[ ! -d \"$ddir\" ]] && mkdir -p \"$ddir\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phred\n",
    "\n",
    "Below is the procedure to compile phred and daev. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-09T23:49:52.504326Z",
     "start_time": "2021-02-09T23:49:50.547986Z"
    }
   },
   "outputs": [],
   "source": [
    "myphred=\"phred-dist-020425.c-acd.tar.Z\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/phred\" ]] && mkdir -p \"$ddir/phred\"\n",
    "tar -C \"$ddir/phred\" -xzvf \"$ddir/$myphred\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:01:10.272424Z",
     "start_time": "2021-02-10T00:01:09.905034Z"
    }
   },
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "\n",
    "cd \"$ddir/phred\"\n",
    "make\n",
    "make daev\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:06:03.844409Z",
     "start_time": "2021-02-10T00:06:03.158024Z"
    }
   },
   "outputs": [],
   "source": [
    "# Copy the executables to the conda environment\n",
    "cp -a \"$ddir/phred/phred\" \"$ddir/phred/daev\" \"$CONDA_PREFIX/bin/\"\n",
    "\n",
    "# Copy the needed file\n",
    "cp -a \"$ddir/phred/phredpar.dat\" \"$CONDA_PREFIX/etc/phredpar.dat\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T04:38:51.147544Z",
     "start_time": "2021-02-10T04:38:50.612284Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Add the machine and chemistry used by the sequencing facility to the phredpar file\n",
    "## Machine model and related information are available in the header of scf files\n",
    "sed -i '/^begin/,/^# Chromatograms/ s/^#$/\"KB_3730_POP7_BDTv3.mob\"        terminator      big-dye                 ABI_3700\\n#/' \"$CONDA_PREFIX/etc/phredpar.dat\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phd2fasta\n",
    "\n",
    "Below is the procedure to compile phd2fasta. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:16:30.229082Z",
     "start_time": "2021-02-10T00:16:29.350941Z"
    }
   },
   "outputs": [],
   "source": [
    "myphd=\"phd2fasta-acd-dist.130911.tar.gz\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/phd\" ]] && mkdir -p \"$ddir/phd\"\n",
    "tar -C \"$ddir/phd\" -xzvf \"$ddir/$myphd\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "\n",
    "cd \"$ddir/phd\"\n",
    "make\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:18:50.120948Z",
     "start_time": "2021-02-10T00:18:49.595683Z"
    }
   },
   "outputs": [],
   "source": [
    "# Copy the executable to the conda environment\n",
    "cp -a \"$ddir/phd/phd2fasta\" \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phrap\n",
    "\n",
    "Below is the procedure to compile phrap. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:26:55.265268Z",
     "start_time": "2021-02-10T00:26:53.920439Z"
    }
   },
   "outputs": [],
   "source": [
    "myphrap=\"distrib.tar.Z\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/phrap.d\" ]] && mkdir -p \"$ddir/phrap.d\"\n",
    "tar -C \"$ddir/phrap.d\" -xzvf \"$ddir/$myphrap\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "\n",
    "cd \"$ddir/phrap.d\"\n",
    "make\n",
    "make manyreads\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:32:17.417847Z",
     "start_time": "2021-02-10T00:32:15.917227Z"
    }
   },
   "outputs": [],
   "source": [
    "# Clean and copy the folder to the conda environment\n",
    "rm \"$ddir/phrap.d/\"*.{c,h,o}\n",
    "cp -aR \"$ddir/phrap.d/\" \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:50:00.792722Z",
     "start_time": "2021-02-10T00:50:00.249473Z"
    }
   },
   "outputs": [],
   "source": [
    "ln -s \"$CONDA_PREFIX/bin/phrap.d/\"{cross_match,cross_match.manyreads,loco,phrap,phrap.longreads,phrap.manyreads,phrapview,swat} \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Consed\n",
    "\n",
    "Below is the procedure to compile consed. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:56:36.857131Z",
     "start_time": "2021-02-10T00:56:19.292381Z"
    }
   },
   "outputs": [],
   "source": [
    "mycs=\"consed_linux.tar.gz\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/consed\" ]] && mkdir -p \"$ddir/consed\"\n",
    "tar -C \"$ddir/consed\" -xzvf \"$ddir/$mycs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T01:11:44.766721Z",
     "start_time": "2021-02-10T01:11:42.787993Z"
    }
   },
   "outputs": [],
   "source": [
    "for i in \"$ddir/consed/\"consed*\n",
    "do\n",
    "    ./\"$i\" -v\n",
    "    [[ $? -eq 0 ]] && prog=$(basename \"$i\") && break\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T01:30:59.586603Z",
     "start_time": "2021-02-10T01:30:58.927395Z"
    }
   },
   "outputs": [],
   "source": [
    "# Correct small bug in install script (when whitespace is present in current path)\n",
    "sed -i 's|cp $szDownloadDirectory|cp \\\\\"$szDownloadDirectory\\\\\"|g' \"$ddir/consed/installConsed.perl\"\n",
    "sed -i 's|cp $szNewDir|cp \\\\\"$szNewDir\\\\\"|g' \"$ddir/consed/installConsed.perl\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "cd \"$ddir/consed/\"\n",
    "export CONSED_HOME=\"$CONDA_PREFIX/bin/consed.d\"\n",
    "mkdir \"$CONSED_HOME\"\n",
    "./installConsed.perl \"$prog\" \"$CONSED_HOME\"\n",
    "cd -\n",
    "ln -s \"$CONDA_PREFIX/bin/consed.d/bin/consed\" \"$CONDA_PREFIX/bin/consed\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### PolyPhred\n",
    "\n",
    "Below is the procedure to compile polyphred. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T02:16:06.914531Z",
     "start_time": "2021-02-10T02:16:05.505317Z"
    }
   },
   "outputs": [],
   "source": [
    "mypp=\"polyphred-6.18-binary-x86_64-unknown-linux-gnu.tar.gz\"\n",
    "\n",
    "## Decompress archive\n",
    "tar -C \"$CONDA_PREFIX/bin/\" -xzvf \"$ddir/$mypp\"\n",
    "\n",
    "# Link executable\n",
    "ln -s \"$CONDA_PREFIX/bin/polyphred-6.18-binary-x86_64-unknown-linux-gnu/bin/\"* \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T02:22:35.275437Z",
     "start_time": "2021-02-10T02:22:34.705677Z"
    }
   },
   "outputs": [],
   "source": [
    "# Correct perl location\n",
    "sed -i '1s|^#!.*|#!/usr/bin/env perl|' \"$CONDA_PREFIX/bin/polyphred-6.18-binary-x86_64-unknown-linux-gnu/bin/phredPhrap.pl\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Environment variable\n",
    "\n",
    "In order to have phred and consed working as expected, two environment variables need to be exported each time a new Jupyter session is run or when the kernel is rebooted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "export PHRED_PARAMETER_FILE=\"$CONDA_PREFIX/etc/phredpar.dat\"\n",
    "export CONSED_HOME=\"$CONDA_PREFIX/bin/consed.d\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequencing data\n",
    "\n",
    "This step downloads the fastq files of the different samples from the SRA repository."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-16T04:48:10.118277Z",
     "start_time": "2021-01-16T04:48:09.890999Z"
    }
   },
   "outputs": [],
   "source": [
    "# Data directory\n",
    "ldir=\"data/libraries\"\n",
    "[[ ! -d \"$ldir\" ]] && mkdir -p \"$ldir\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-16T04:52:28.344316Z",
     "start_time": "2021-01-16T04:52:23.090603Z"
    }
   },
   "outputs": [],
   "source": [
    "# Bioproject\n",
    "bioproject=ERP114942\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Keep S. mansoni samples only\n",
    "fds=$(head -1 runinfo | tr \",\" \"\\n\" | grep -w -n \"ScientificName\" | cut -d \":\" -f 1)\n",
    "awk -v fds=$fds -F \",\" '$fds == \"Schistosoma mansoni\"' runinfo > runinfo_sm\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir/$fln/\" ]] && mkdir -p \"$ldir/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: dowloading problem\" >> \"$ldir/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir/$fln/\" --split-files \"$ldir/$fln/$run\"\n",
    "    rm \"$ldir/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir/$fln/${run}_1.fastq\" \"$ldir/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir/$fln/${run}_2.fastq\" \"$ldir/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < runinfo_sm\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir/\"*/*\n",
    "\n",
    "rm runinfo*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download our exome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-01-15T23:21:41.162Z"
    }
   },
   "outputs": [],
   "source": [
    "# Bioproject\n",
    "bioproject=PRJNA560069\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir/$fln/\" ]] && mkdir -p \"$ldir/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: dowloading problem\" >> \"$ldir/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir/$fln/\" --split-files \"$ldir/$fln/$run\"\n",
    "    rm \"$ldir/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir/$fln/${run}_1.fastq\" \"$ldir/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir/$fln/${run}_2.fastq\" \"$ldir/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < <(tail -n +2 runinfo | sed \"/^$/d\")\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir/\"*/*\n",
    "\n",
    "rm runinfo*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genome data\n",
    "\n",
    "The genome data is downloaded from the [WormBase ParaSite](https://parasite.wormbase.org). We use the data from the version 14 (WBPS14). The data is then indexed for the different tools used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-11T01:12:15.279700Z",
     "start_time": "2021-01-11T01:12:15.122045Z"
    }
   },
   "outputs": [],
   "source": [
    "gdir=\"data/genome\"\n",
    "[[ ! -d \"$gdir\" ]] && mkdir -p \"$gdir\"\n",
    "\n",
    "# Download and unzip data\n",
    "wget -P \"$gdir\" ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS14/species/schistosoma_mansoni/PRJEA36577/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa.gz\n",
    "pigz -d \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa.gz\"\n",
    "\n",
    "# Preparing indices\n",
    "bwa index \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "samtools faidx \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "gatk CreateSequenceDictionary -R \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "\n",
    "# Download the S. mansoni genome annotation\n",
    "wget -P \"$gdir\" ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS14/species/schistosoma_mansoni/PRJEA36577/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3.gz\n",
    "pigz -d \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Known variants\n",
    "\n",
    "List of variants from the SmLE x SmHR crosses that segregated in a Mendelian fashion in F1 ([Valentim *et al.* 2013](https://doi.org/10.1126/science.1243106)). This list was generated from alignments using the v5 genome. This has been lifted over to the latest genome using the [flo pipeline](https://github.com/wurmlab/flo/tree/727f10b2b1c57a0514835d302d7f6345d3a34ffb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-11T04:04:26.444016Z",
     "start_time": "2021-01-11T04:04:20.050912Z"
    }
   },
   "outputs": [],
   "source": [
    "# Index file\n",
    "gatk IndexFeatureFile -I \"$gdir/sm_dbSNP_v7.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-14T19:41:52.029609Z",
     "start_time": "2021-01-14T19:41:51.708951Z"
    }
   },
   "outputs": [],
   "source": [
    "# For Freebayes\n",
    "bgzip -c \"$gdir/sm_dbSNP_v7.vcf\" > \"$gdir/sm_dbSNP_v7.vcf.gz\"\n",
    "tabix \"$gdir/sm_dbSNP_v7.vcf.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T04:54:05.168847Z",
     "start_time": "2021-01-21T04:52:32.583105Z"
    }
   },
   "outputs": [],
   "source": [
    "find data/libraries/ -type d -name \"*SAM*\"   | sed \"s|.*/||g\" | sed \"s|$|\\tUganda|g\" > data/pop\n",
    "find data/libraries/ -type d -name \"*Sm.TZ*\" | sed \"s|.*/||g\" | sed \"s|$|\\tTanzania|g\" >> data/pop\n",
    "find data/libraries/ -type d -name \"*Sm.SN*\" | sed \"s|.*/||g\" | sed \"s|$|\\tSenegal|g\" >> data/pop\n",
    "find data/libraries/ -type d -name \"*Sm.NE*\" | sed \"s|.*/||g\" | sed \"s|$|\\tNiger|g\" >> data/pop\n",
    "find data/libraries/ -type d -name \"*Sm.BR*\" | sed \"s|.*/||g\" | sed \"s|$|\\tBrazil|g\" >> data/pop\n",
    "find data/libraries/ -type d -name \"*Sm.OM*\" | sed \"s|.*/||g\" | sed \"s|$|\\tOman|g\" >> data/pop\n",
    "#find data/libraries/ -type d -name *Sr* | sed \"s|.*/||g\" | sed \"s|$|\\tSro|g\" >> data/pop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GFF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-20T04:39:09.414147Z",
     "start_time": "2021-01-20T04:39:08.235548Z"
    }
   },
   "outputs": [],
   "source": [
    "# Gene code\n",
    "mygene=Smp_246790\n",
    "grep \"$mygene\" \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3\" > \"$gdir/$mygene.gff\"\n",
    "\n",
    "# Generate isoform specific gff\n",
    "myiso=$(grep -o \"$mygene\\.[0-9]*\" \"$gdir/$mygene.gff\" | sort | uniq)\n",
    "\n",
    "for i in $myiso\n",
    "do\n",
    "    awk '$3 == \"gene\"' \"$gdir/$mygene.gff\" > \"$gdir/$i.gff\"\n",
    "    grep -w \"$i\" \"$gdir/$mygene.gff\" | grep -v \"exon\" >> \"$gdir/$i.gff\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequencing data processing\n",
    "\n",
    "### Snakemake pipeline\n",
    "\n",
    "The snakemake pipeline performs the following steps:\n",
    "* alignment of each library against the reference genome\n",
    "* marking duplicates\n",
    "* base quality recalibration\n",
    "* generating alignment statistics\n",
    "* outputing Smp_246790 gene read depth for each library\n",
    "* outputing Z chromosome read depth for each library\n",
    "* determining sex of each individual\n",
    "* calling variants (gVCF) for each library\n",
    "* combining gVCF of all libraries\n",
    "* genotyping all samples per contig\n",
    "* merging all the contigs to generate final VCF file\n",
    "\n",
    "The VCF file produced will then be used to identify possible QTLs link to the cercarial production."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-20T03:57:03.036418Z",
     "start_time": "2021-01-20T03:57:02.821552Z"
    }
   },
   "outputs": [],
   "source": [
    "# Status directory\n",
    "statdir=status\n",
    "[[ ! -d \"$statdir\" ]] && mkdir \"$statdir\"\n",
    "\n",
    "snakemake --snakefile snakefile --cluster \"qsub -V -cwd -o $statdir -j y -r y -pe smp 10 -S /bin/bash\" --jobs 1000 -w 120"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant calling\n",
    "\n",
    "We call the variants within the gene region across all samples directly from the BAM files using FreeBayes. This method is preferred because it can perform population defined calling."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T05:06:16.102298Z",
     "start_time": "2021-01-21T04:56:00.555221Z"
    }
   },
   "outputs": [],
   "source": [
    "# Output directory\n",
    "cdir=\"data/variant_calling\"\n",
    "[[ ! -d \"$cdir\" ]] && mkdir -p \"$cdir\"\n",
    "\n",
    "# Variables\n",
    "mygenome=\"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "#mypos=SM_V7_3:600000-1000000\n",
    "mypos=SM_V7_3:630000-800000\n",
    "\n",
    "myfiles=$(find data -name *SAM*.bam -o -name *Sm*.bam | sort | tr \"\\n\" \" \") ## !! TO CHANGE SAMPLES\n",
    "\n",
    "myfilename=\"$cdir/PZQ-R_field\"\n",
    "\n",
    "\n",
    "# Command used for calling variants\n",
    "freebayes -f \"$mygenome\" \\\n",
    "   -r \"$mypos\" \\\n",
    "   -b $myfiles \\\n",
    "   -=          \\\n",
    "   --population data/pop \\\n",
    "   -q 20       \\\n",
    "   -m 30       \\\n",
    "   -@ \"$gdir/sm_dbSNP_v7.vcf.gz\" > \"${myfilename}.vcf\"\n",
    "   \n",
    "#     -F 0.4      \\\n",
    "\n",
    "pigz \"${myfilename}\".vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyzing $SmTRP_{PZQ}$\n",
    "\n",
    "We will analyze the previously called variants on the gene only. This analysis includes:\n",
    "* Filtering out genotypes with low read depth,\n",
    "* Removing invariable sites,\n",
    "* Normalizing coordinates,\n",
    "* Use a custom script to summarize variants and genotype from the VCF file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T05:16:58.645883Z",
     "start_time": "2021-01-21T05:13:33.504609Z"
    }
   },
   "outputs": [],
   "source": [
    "# Remove genotypes with low read depths\n",
    "vcftools --gzvcf \"${myfilename}.vcf.gz\" \\\n",
    "    --minDP 4 \\\n",
    "    --recode  \\\n",
    "    --recode-INFO-all \\\n",
    "    --out \"${myfilename}.flt-dp4\"\n",
    "\n",
    "# Rename file\n",
    "mv \"${myfilename}.flt-dp4.recode.vcf\" \"${myfilename}.flt-dp4.vcf\"\n",
    "\n",
    "# Clean the VCF of invariable ref variants\n",
    "vcf-subset -a -e \"${myfilename}.flt-dp4.vcf\" > \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "\n",
    "# # Normalize VCF\n",
    "# vt normalize -o \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "#     -r \"$mygenome\" \\\n",
    "#     \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "## This fails with a core-dump\n",
    "\n",
    "## Maybe look at bcftools norm (with -f option)   !!! Install bcftools\n",
    "bcftools norm -o \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "    -f \"$mygenome\" \\\n",
    "    \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "\n",
    "##Another method of normalization\n",
    "# vcfleftalign -r \"$mygenome\" \\\n",
    "#     \"${myfilename}.flt-dp4.cleaned.vcf\" > \"${myfilename}.flt-dp4.cleaned.norm.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T05:20:42.965528Z",
     "start_time": "2021-01-21T05:20:41.291224Z"
    }
   },
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir=\"results/1-reports\"\n",
    "[[ ! -d \"$rdir\" ]] && mkdir -p \"$rdir\"\n",
    "\n",
    "# Generate reports per isoform\n",
    "for i in $myiso\n",
    "do\n",
    "    qsub -V -cwd -o \"$statdir\" -j y -b y scripts/func-table-mut.sh -v \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "        -r \"$mygenome\" \\\n",
    "        -g \"$gdir/$i.gff\" \\\n",
    "        -p data/pop \\\n",
    "        -o \"$rdir/$i.flt.norm.tsv\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global coverage of the gene\n",
    "\n",
    "We inspect how well the gene and particularly the exons are covered. This is done by concatenating individual coverage of the gene and then plotting total read depth at each position. The concatenation step can take ~20 minutes or more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-23T01:04:45.700250Z",
     "start_time": "2021-01-23T00:46:02.886902Z"
    }
   },
   "outputs": [],
   "source": [
    "fl_out=f\n",
    "\n",
    "fl_ls=($(find data/libraries/ -type f -name \"*PZQ.cov\" | sort))\n",
    "\n",
    "# Process first file\n",
    "echo -e \"#CHROM\\t\"$(cut -f1 \"${fl_ls[0]}\" | tr \"\\n\" \"\\t\" | sed \"s/\\t$//\") > \"$fl_out\"\n",
    "echo -e \"POS\\t\"$(cut -f2 \"${fl_ls[0]}\"| tr \"\\n\" \"\\t\" | sed \"s/\\t$//\") > \"$fl_out\"\n",
    "\n",
    "for ((i = 0 ; i < ${#fl_ls[@]} ; i++))\n",
    "do\n",
    "echo $i\n",
    "    echo -e $(dirname \"${fl_ls[$i]}\" | sed \"s,.*/,,\")\"\\t\"$(cut -f 3 \"${fl_ls[$i]}\" | tr \"\\n\" \"\\t\" | sed \"s/\\t$//\")  > \"$fl_out\"\n",
    "done"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Plotting data\n",
    "\n",
    "Rscript to be written"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutations of interest\n",
    "\n",
    "Some of the mutations are of high interest because located in critical sections of the protein (PZQ binding pocket, channel, TRP box). We list the samples that carry them in order to look at the reads within these samples to assess the likelihood of these mutations. This will also help identify samples that will require further investigation like PCR validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-25T17:55:43.118997Z",
     "start_time": "2021-01-25T17:54:48.030156Z"
    }
   },
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir2=\"results/2-mutations of interest\"\n",
    "[[ ! -d \"$rdir2\" ]] && mkdir -p \"$rdir2\"\n",
    "\n",
    "# mutations=\"p.T1394I p.Q1432H p.S1448Y p.G1458C p.H1522Q p.I1523N p.P1527T p.P1527Q p.Q1598K p.D1602Y p.D1606Y p.Q1609L p.Q1673K p.M1674I p.D1677Y p.H1680N p.P1683Q p.P1683R p.L1684I p.P1686T p.P1686Q p.P1687Q p.W1692L p.E1696* p.A1700D p.Q1704K\"\n",
    "mutations=\"p.H1522Q p.I1523N p.P1527T p.P1527Q p.Q1598K p.D1602Y p.D1606Y p.Q1609L p.Q1673K p.M1674I p.D1677Y p.H1680N p.P1686T p.P1686Q p.P1687Q\"\n",
    "\n",
    "# Add stop codon\n",
    "mutations+=\" $(awk '$8 ~ /*/ {print $8}' $rdir/Smp_246790.5.flt.norm.tsv)\"\n",
    "\n",
    "# Isoform of interest\n",
    "myiso=Smp_246790.5\n",
    "\n",
    "unset mylist\n",
    "for m in $mutations\n",
    "do\n",
    "echo $m\n",
    "    mypos=$(awk -v m=$m '$8 == m {print $1}' \"$rdir/$myiso.flt.norm.tsv\")\n",
    "    myspl=$(awk -v mypos=$mypos '$2 == mypos' \"${myfilename}.flt-dp4.cleaned.norm.vcf\"| tr \"\\t\" \"\\n\" | grep -n 0/1 | cut -d \":\" -f 1 | tr \"\\n\" \",\" | sed \"s/,$//\")\n",
    "    myspl=$(grep -m 1 \"#C\" \"${myfilename}.flt-dp4.cleaned.norm.vcf\" | cut -f $myspl)\n",
    "    mylist+=\"$m\\t$mypos\\t$myspl\"\"\\n\" \n",
    "done\n",
    "\n",
    "echo -ne \"$mylist\" > \"$rdir2/samples_list.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output read depth for each sample carrying mutations of interest\n",
    "Rscript scripts/variant_read_depth.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plots of the read depth for each mutation and sample are available in the folder labeled graphs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Primer design\n",
    "\n",
    "Interesting mutations from the calling data must be checked by Sanger sequencing. These mutations are located in specific exons. To sequence the full exons, we extract exon sequences with an additional 200 bp on each end, where primers should be designed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-09T20:04:08.658360Z",
     "start_time": "2021-02-09T20:04:02.081894Z"
    }
   },
   "outputs": [],
   "source": [
    "fl_out=\"$rdir2/exons_of_interest\"\n",
    "\n",
    "myexons=(3 4 12 23 25 27 29 34)\n",
    "myiso=\"Smp_246790.5\"\n",
    "\n",
    "grep \"$myiso\" \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3\" | \\\n",
    "    awk '$3 == \"CDS\"'  | \\\n",
    "    sed -n \"$(sed \"s/ /p;/g\" <<< ${myexons[@]})p\" | \\\n",
    "    awk '{print $1\"\\t\"$4-201\"\\t\"$5+200}' > \"$fl_out.bed\"\n",
    "\n",
    "bedtools getfasta -fi \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\" \\\n",
    "    -bed \"$fl_out.bed\" \\\n",
    "    -fo \"$fl_out.fas\"\n",
    "\n",
    "# Reformat fasta entry\n",
    "myentries=($(grep -n \">\" \"$fl_out.fas\" | cut -d \":\" -f 1))\n",
    "for ((i = 0 ; i < \"${#myexons[@]}\" ; i++))\n",
    "do\n",
    "    sed -i \"${myentries[$i]}s/>/>E${myexons[$i]} /\" \"$fl_out.fas\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sanger sequencing\n",
    "\n",
    "Sanger sequencing traces are analyzed using Consed and PolyPhred: The first align the data while the second perform the calling. PolyPhred output is then processed with a custom script in order to summarize the calling per sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data folder\n",
    "mkdir data/sanger_sequencing\n",
    "\n",
    "# Download the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:11:18.003687Z",
     "start_time": "2021-02-07T06:11:15.072648Z"
    }
   },
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir3=\"results/2-mutations of interest/sanger_sequencing\"\n",
    "[[ ! -d \"$rdir3\" ]] && mkdir -p \"$rdir3\"\n",
    "\n",
    "# Phred folders\n",
    "mkdir -p \"$rdir3/data/\"{chromat_dir,edit_dir,phd_dir,poly_dir}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:11:18.003687Z",
     "start_time": "2021-02-07T06:11:15.072648Z"
    }
   },
   "outputs": [],
   "source": [
    "mygene=\"Smp_246790\"\n",
    "\n",
    "# Generate reference sequence\n",
    "mkdir \"$rdir3/0-Ref_seq\"\n",
    "bedtools getfasta -fi \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\" \\\n",
    "    -bed <(grep \"$mygene\" \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3\" | awk '$3 == \"gene\"' | awk '{print $1\"\\t\"$4-1\"\\t\"$5}') \\\n",
    "    -fo \"$rdir3/Smp_246790.fas\"\n",
    "\n",
    "# Rename fasta entry\n",
    "sed -i \"s/>.*/>$mygene/\" \"$rdir3/Smp_246790.fas\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:15:44.258775Z",
     "start_time": "2021-02-07T06:15:38.876037Z"
    }
   },
   "outputs": [],
   "source": [
    "pwd_old=$PWD\n",
    "\n",
    "# Generate ref sequence for phred\n",
    "cd \"$rdir3/\"\n",
    "sudophred \"$mygene.fas\" -r -abi\n",
    "\n",
    "# Move ref\n",
    "mv \"$mygene.REF\" \"data/chromat_dir/\"\n",
    "mv \"$mygene.REF.phd.1\" \"data/phd_dir/\"\n",
    "\n",
    "cd \"$pwd_old\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare data\n",
    "\n",
    "We create links to the original data but with modified file names in order to informed Polyphred about the paired samples. Samples will be identified on the first 8 characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:27:12.448693Z",
     "start_time": "2021-02-07T06:23:27.111093Z"
    }
   },
   "outputs": [],
   "source": [
    "# Generate modified sample names for polyphred\n",
    "j=1\n",
    "unset spl_code\n",
    "for i in $(tail -n +2 \"$rdir3/barcode_list.tsv\" | cut -f 1 | sort | uniq)\n",
    "do\n",
    "    spl_code+=\"$i\\t$(sed \"s/^/$(printf \"%02d\" $j)_/\" <<< \"$i\")\\n\"\n",
    "    ((j++))\n",
    "done\n",
    "\n",
    "# Generate exon specific folder and link to data\n",
    "while read line \n",
    "do\n",
    "    # Sample name\n",
    "    myspl=$(cut -f 1 <<< \"$line\")\n",
    "    myspl=$(echo -ne \"$spl_code\" | grep \"$myspl\" | cut -f2)\n",
    "    \n",
    "    # Barcode\n",
    "    mycode=$(cut -f 3 <<< \"$line\")\n",
    "    [[ \"$mycode\" == NA ]] && continue\n",
    "    \n",
    "    # Link files\n",
    "    ln -s \"$PWD/data/sanger_sequencing/$mycode\"*.scf \"$rdir3/data/chromat_dir/${myspl}_$mycode.scf\"\n",
    "\n",
    "done < <(tail -n +2 \"$rdir3/barcode_list.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-09T21:46:40.756484Z",
     "start_time": "2021-02-09T21:46:40.238240Z"
    }
   },
   "source": [
    "Run Phred and Consed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:28:57.536280Z",
     "start_time": "2021-02-07T06:28:30.561909Z"
    }
   },
   "outputs": [],
   "source": [
    "pwd_old=$PWD\n",
    "\n",
    "export PHRED_PARAMETER_FILE=/master/fcheval/local/bin/consed_29.0/lib/phredpar.dat\n",
    "\n",
    "# Moving to the working folder\n",
    "cd \"$rdir3/data/edit_dir\"\n",
    "\n",
    "# Generating phd and poly files\n",
    "phred ../chromat_dir/*.scf -pd ../phd_dir/ -dd ../poly_dir/\n",
    "\n",
    "# Preparing the reference sequence\n",
    "phd2Ace.perl \"$mygene.REF.phd.1\"\n",
    "\n",
    "# Aligning reads to the reference\n",
    "find ../chromat_dir/* -name \"*.scf\" -exec basename {} \\; > reads_to_add.fof\n",
    "consed -ace \"$mygene.REF.ace\" -addNewReads reads_to_add.fof -newAceFilename \"$mygene.REF.ace.1\"\n",
    "cd \"$pwd_old\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run PolyPhred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:32:37.758770Z",
     "start_time": "2021-02-07T06:32:07.204807Z"
    }
   },
   "outputs": [],
   "source": [
    "score_snp=70 #90\n",
    "score_indel=70 #80\n",
    "qual=25 #40\n",
    "\n",
    "# SNP scoring\n",
    "name=$mygene.s${score_snp}-q$qual\n",
    "polyphred -d \"$rdir3/data\" -extended_genotype -n $name.snp.nav -o $name.snp.polyphred.out -q $qual -refcomp -s 1 8 -score $score_snp\n",
    "\n",
    "# Indel scoring\n",
    "name=$mygene.s${score_indel}-q$qual\n",
    "polyphred -d \"$rdir3/data\" -extended_genotype -i -inav $name.indel.nav -o $name.indel.polyphred.out -q $qual -s 1 8 -iscore $score_indel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:33:19.572311Z",
     "start_time": "2021-02-07T06:33:18.502676Z"
    }
   },
   "outputs": [],
   "source": [
    "# Genotype extractions\n",
    "scripts/trim_polyphred.sh -i \"$rdir3/data/edit_dir/$mygene.s${score_snp}-q$qual.snp.polyphred.out\" \\\n",
    "    -o \"$rdir3/$mygene.SNP.gt.tsv\" \\\n",
    "    -p 2 3 \\\n",
    "    -d \"_\"\n",
    "\n",
    "scripts/trim_polyphred.sh -i \"$rdir3/data/edit_dir/$mygene.s${score_indel}-q$qual.indel.polyphred.out\" \\\n",
    "    -o \"$rdir3/$mygene.INDEL.gt.tsv\" \\\n",
    "    -p 2 3 \\\n",
    "    -d \"_\"\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "384px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
