{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PZQ-R field variants\n",
    "\n",
    "\n",
    "We score and investigate variants from natural populations in order to have a comprehensive view of those that could impact praziquantel efficacy. We used data from different locations in New and Old World:\n",
    "* Brazil (original from the associated publication)\n",
    "* Senegal\n",
    "* Niger\n",
    "* Tanzania\n",
    "* Uganda (data generated by Sanger https://doi.org/10.3389/fgene.2019.00826)\n",
    "* Oman\n",
    "\n",
    "\n",
    "## Environment and data\n",
    "\n",
    "### Environment\n",
    "\n",
    "Creating a conda environment improves reproducibility by installing specific versions of the programs used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-01-06T16:09:29.118Z"
    }
   },
   "outputs": [],
   "source": [
    "conda env create -f .env/env.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cell below must be run each time a new Jupyter session is run or when the kernel is rebooted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-08T19:25:05.636562Z",
     "start_time": "2021-06-08T19:25:00.717400Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Activate the environment\n",
    "source $(sed \"s,/bin/conda,,\" <<<$CONDA_EXE)/etc/profile.d/conda.sh\n",
    "conda activate PZQ-R_field\n",
    "\n",
    "# Remove potential variable interferences\n",
    "export PERL5LIB=\"\"\n",
    "export PYTHONNOUSERSITE=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cell below must be run only once at the time of the environment creation. The R script install the R dependencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T22:49:52.915105Z",
     "start_time": "2021-01-21T22:49:13.794595Z"
    }
   },
   "outputs": [],
   "source": [
    "# Installing needed R packages\n",
    "Rscript \".env/R package dependencies.R\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Phred/Phrap/Consed and PolyPhred\n",
    "\n",
    "[Phred/Phrap/Consed](http://www.phrap.org/phredphrapconsed.html) and [PolyPhred](https://droog.gs.washington.edu/polyphred/) are software used to analyze Sanger sequencing traces by performing alignment and genotyping, respectively. They are freely available for academic used. However, this requires to contact the authors in order to get a copy ([here](http://www.phrap.org/consed/consed.html#howToGet) for Phred/Phrap/Consed, [here](https://droog.gs.washington.edu/polyphred/poly_get.html) for PolyPhred).\n",
    "\n",
    "Once the copies obtained, they must be place in a download folder at the same level as this notebook. Then programs can be compiled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-09T23:49:34.438439Z",
     "start_time": "2021-02-09T23:49:34.224809Z"
    }
   },
   "outputs": [],
   "source": [
    "# Folder in which the software archives must be located\n",
    "ddir=\"download\"\n",
    "[[ ! -d \"$ddir\" ]] && mkdir -p \"$ddir\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phred\n",
    "\n",
    "Below is the procedure to compile phred and daev. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-09T23:49:52.504326Z",
     "start_time": "2021-02-09T23:49:50.547986Z"
    }
   },
   "outputs": [],
   "source": [
    "myphred=\"phred-dist-020425.c-acd.tar.Z\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/phred\" ]] && mkdir -p \"$ddir/phred\"\n",
    "tar -C \"$ddir/phred\" -xzvf \"$ddir/$myphred\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:01:10.272424Z",
     "start_time": "2021-02-10T00:01:09.905034Z"
    }
   },
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "\n",
    "cd \"$ddir/phred\"\n",
    "make\n",
    "make daev\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:06:03.844409Z",
     "start_time": "2021-02-10T00:06:03.158024Z"
    }
   },
   "outputs": [],
   "source": [
    "# Copy the executables to the conda environment\n",
    "cp -a \"$ddir/phred/phred\" \"$ddir/phred/daev\" \"$CONDA_PREFIX/bin/\"\n",
    "\n",
    "# Copy the needed file\n",
    "cp -a \"$ddir/phred/phredpar.dat\" \"$CONDA_PREFIX/etc/phredpar.dat\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T04:38:51.147544Z",
     "start_time": "2021-02-10T04:38:50.612284Z"
    }
   },
   "outputs": [],
   "source": [
    "# Add the machine and chemistry used by the sequencing facility to the phredpar file\n",
    "## Machine model and related information are available in the header of scf files\n",
    "sed -i '/^begin/,/^# Chromatograms/ s/^#$/\"KB_3730_POP7_BDTv3.mob\"        terminator      big-dye                 ABI_3700\\n#/' \"$CONDA_PREFIX/etc/phredpar.dat\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phd2fasta\n",
    "\n",
    "Below is the procedure to compile phd2fasta. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:16:30.229082Z",
     "start_time": "2021-02-10T00:16:29.350941Z"
    }
   },
   "outputs": [],
   "source": [
    "myphd=\"phd2fasta-acd-dist.130911.tar.gz\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/phd\" ]] && mkdir -p \"$ddir/phd\"\n",
    "tar -C \"$ddir/phd\" -xzvf \"$ddir/$myphd\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "\n",
    "cd \"$ddir/phd\"\n",
    "make\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:18:50.120948Z",
     "start_time": "2021-02-10T00:18:49.595683Z"
    }
   },
   "outputs": [],
   "source": [
    "# Copy the executable to the conda environment\n",
    "cp -a \"$ddir/phd/phd2fasta\" \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phrap\n",
    "\n",
    "Below is the procedure to compile phrap. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:26:55.265268Z",
     "start_time": "2021-02-10T00:26:53.920439Z"
    }
   },
   "outputs": [],
   "source": [
    "myphrap=\"distrib.tar.Z\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/phrap.d\" ]] && mkdir -p \"$ddir/phrap.d\"\n",
    "tar -C \"$ddir/phrap.d\" -xzvf \"$ddir/$myphrap\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "\n",
    "cd \"$ddir/phrap.d\"\n",
    "make\n",
    "make manyreads\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:32:17.417847Z",
     "start_time": "2021-02-10T00:32:15.917227Z"
    }
   },
   "outputs": [],
   "source": [
    "# Clean and copy the folder to the conda environment\n",
    "rm \"$ddir/phrap.d/\"*.{c,h,o}\n",
    "cp -aR \"$ddir/phrap.d/\" \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:50:00.792722Z",
     "start_time": "2021-02-10T00:50:00.249473Z"
    }
   },
   "outputs": [],
   "source": [
    "ln -s \"$CONDA_PREFIX/bin/phrap.d/\"{cross_match,cross_match.manyreads,loco,phrap,phrap.longreads,phrap.manyreads,phrapview,swat} \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Consed\n",
    "\n",
    "Below is the procedure to compile consed. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T00:56:36.857131Z",
     "start_time": "2021-02-10T00:56:19.292381Z"
    }
   },
   "outputs": [],
   "source": [
    "mycs=\"consed_linux.tar.gz\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/consed\" ]] && mkdir -p \"$ddir/consed\"\n",
    "tar -C \"$ddir/consed\" -xzvf \"$ddir/$mycs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T01:11:44.766721Z",
     "start_time": "2021-02-10T01:11:42.787993Z"
    }
   },
   "outputs": [],
   "source": [
    "for i in \"$ddir/consed/\"consed*\n",
    "do\n",
    "    ./\"$i\" -v\n",
    "    [[ $? -eq 0 ]] && prog=$(basename \"$i\") && break\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T01:30:59.586603Z",
     "start_time": "2021-02-10T01:30:58.927395Z"
    }
   },
   "outputs": [],
   "source": [
    "# Correct small bug in install script (when whitespace is present in current path)\n",
    "sed -i 's|cp $szDownloadDirectory|cp \\\\\"$szDownloadDirectory\\\\\"|g' \"$ddir/consed/installConsed.perl\"\n",
    "sed -i 's|cp $szNewDir|cp \\\\\"$szNewDir\\\\\"|g' \"$ddir/consed/installConsed.perl\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "cd \"$ddir/consed/\"\n",
    "export CONSED_HOME=\"$CONDA_PREFIX/bin/consed.d\"\n",
    "mkdir \"$CONSED_HOME\"\n",
    "./installConsed.perl \"$prog\" \"$CONSED_HOME\"\n",
    "cd -\n",
    "ln -s \"$CONDA_PREFIX/bin/consed.d/bin/consed\" \"$CONDA_PREFIX/bin/consed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-05T20:58:59.619Z"
    }
   },
   "outputs": [],
   "source": [
    "ln -s \"$CONDA_PREFIX/bin/phrap.d/\"{cross_match,cross_match.manyreads} \"$CONDA_PREFIX/bin/consed.d/bin\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-05T21:03:59.421721Z",
     "start_time": "2021-05-05T21:03:59.226966Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "ln -s \"$CONDA_PREFIX/bin/phred\" \"$CONDA_PREFIX/bin/consed.d/bin\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-05T21:10:39.042492Z",
     "start_time": "2021-05-05T21:10:38.160503Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "ln -s \"$CONDA_PREFIX/etc/phredpar.dat\" \"$CONDA_PREFIX/bin/consed.d/lib\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### PolyPhred\n",
    "\n",
    "Below is the procedure to compile polyphred. This requires to have access to a terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T02:16:06.914531Z",
     "start_time": "2021-02-10T02:16:05.505317Z"
    }
   },
   "outputs": [],
   "source": [
    "mypp=\"polyphred-6.18-binary-x86_64-unknown-linux-gnu.tar.gz\"\n",
    "\n",
    "## Decompress archive\n",
    "tar -C \"$CONDA_PREFIX/bin/\" -xzvf \"$ddir/$mypp\"\n",
    "\n",
    "# Link executable\n",
    "ln -s \"$CONDA_PREFIX/bin/polyphred-6.18-binary-x86_64-unknown-linux-gnu/bin/\"* \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-10T02:22:35.275437Z",
     "start_time": "2021-02-10T02:22:34.705677Z"
    }
   },
   "outputs": [],
   "source": [
    "# Correct perl location\n",
    "sed -i '1s|^#!.*|#!/usr/bin/env perl|' \"$CONDA_PREFIX/bin/polyphred-6.18-binary-x86_64-unknown-linux-gnu/bin/phredPhrap.pl\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Environment variable\n",
    "\n",
    "In order to have phred and consed working as expected, two environment variables need to be exported each time a new Jupyter session is run or when the kernel is rebooted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T17:48:34.320937Z",
     "start_time": "2021-05-20T17:48:34.160726Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "export PHRED_PARAMETER_FILE=\"$CONDA_PREFIX/etc/phredpar.dat\"\n",
    "export CONSED_HOME=\"$CONDA_PREFIX/bin/consed.d\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequencing data\n",
    "\n",
    "This step downloads the fastq files of the different samples from the SRA repository."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-25T15:27:14.985455Z",
     "start_time": "2021-05-25T15:27:14.353336Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Data directory\n",
    "ldir=\"data/libraries\"\n",
    "[[ ! -d \"$ldir\" ]] && mkdir -p \"$ldir\"\n",
    "\n",
    "# Data directory for S. mansoni\n",
    "ldir_sm=\"$ldir/1-Sm\"\n",
    "[[ ! -d \"$ldir_sm\" ]] && mkdir -p \"$ldir_sm\"\n",
    "\n",
    "# Data directory for S. haematobium\n",
    "ldir_sh=\"$ldir/2-Sh\"\n",
    "[[ ! -d \"$ldir_sh\" ]] && mkdir -p \"$ldir_sh\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-16T04:52:28.344316Z",
     "start_time": "2021-01-16T04:52:23.090603Z"
    }
   },
   "outputs": [],
   "source": [
    "# Bioproject\n",
    "bioproject=ERP114942\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Keep S. mansoni samples only\n",
    "fds=$(head -1 runinfo | tr \",\" \"\\n\" | grep -w -n \"ScientificName\" | cut -d \":\" -f 1)\n",
    "awk -v fds=$fds -F \",\" '$fds == \"Schistosoma mansoni\"' runinfo > runinfo_sm\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir_sm/$fln/\" ]] && mkdir -p \"$ldir_sm/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir_sm/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir_sm/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: dowloading problem\" >> \"$ldir_sm/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir_sm/$fln/\" --split-files \"$ldir_sm/$fln/$run\"\n",
    "    rm \"$ldir_sm/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir_sm/$fln/${run}_1.fastq\" \"$ldir_sm/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir_sm/$fln/${run}_2.fastq\" \"$ldir_sm/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < runinfo_sm\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir_sm/\"*/*\n",
    "\n",
    "rm runinfo*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download our exome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-01-15T23:21:41.162Z"
    }
   },
   "outputs": [],
   "source": [
    "# Bioproject\n",
    "bioproject=PRJNA560069\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir/$fln/\" ]] && mkdir -p \"$ldir/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: dowloading problem\" >> \"$ldir/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir/$fln/\" --split-files \"$ldir/$fln/$run\"\n",
    "    rm \"$ldir/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir/$fln/${run}_1.fastq\" \"$ldir/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir/$fln/${run}_2.fastq\" \"$ldir/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < <(tail -n +2 runinfo | sed \"/^$/d\")\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir/\"*/*\n",
    "\n",
    "rm runinfo*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*S. haematobium* data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-13T01:04:16.171769Z",
     "start_time": "2021-02-13T01:04:14.853683Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Bioproject\n",
    "bioproject=PRJNA443709\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir_sh/$fln/\" ]] && mkdir -p \"$ldir_sh/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir_sh/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir_sh/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: dowloading problem\" >> \"$ldir_sh/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir_sh/$fln/\" --split-files \"$ldir_sh/$fln/$run\"\n",
    "    rm \"$ldir_sh/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir_sh/$fln/${run}_1.fastq\" \"$ldir_sh/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir_sh/$fln/${run}_2.fastq\" \"$ldir_sh/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < <(tail -n +2 runinfo | sed \"/^$/d\")\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir_sh/\"*/*\n",
    "\n",
    "rm runinfo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genome data\n",
    "\n",
    "The genome data is downloaded from the [WormBase ParaSite](https://parasite.wormbase.org). We use the data from the version 14 (WBPS14) for *S. mansoni* and version 15 (WBPS14) for *S. haematobium*. The data is then indexed for the different tools used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-25T15:26:33.044564Z",
     "start_time": "2021-05-25T15:26:32.802889Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "gdir=\"data/genome\"\n",
    "[[ ! -d \"$gdir\" ]] && mkdir -p \"$gdir\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-11T01:12:15.279700Z",
     "start_time": "2021-01-11T01:12:15.122045Z"
    }
   },
   "outputs": [],
   "source": [
    "# Download and unzip data\n",
    "wget -P \"$gdir\" ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS14/species/schistosoma_mansoni/PRJEA36577/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa.gz\n",
    "pigz -d \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa.gz\"\n",
    "\n",
    "# Preparing indices\n",
    "bwa index \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "samtools faidx \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "gatk CreateSequenceDictionary -R \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "\n",
    "# Download the S. mansoni genome annotation\n",
    "wget -P \"$gdir\" ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS14/species/schistosoma_mansoni/PRJEA36577/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3.gz\n",
    "pigz -d \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-12T22:52:07.989626Z",
     "start_time": "2021-02-12T22:36:31.770043Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) --2021-02-12 16:36:32--  ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS15/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa.gz\n",
      "           => ‘data/genome/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa.gz’\n",
      "Resolving ftp.ebi.ac.uk... 193.62.193.138\n",
      "Connecting to ftp.ebi.ac.uk|193.62.193.138|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/databases/wormbase/parasite/releases/WBPS15/species/schistosoma_haematobium/PRJNA78265 ... done.\n",
      "==> SIZE schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa.gz ... 113732870\n",
      "==> PASV ... done.    ==> RETR schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa.gz ... done.\n",
      "Length: 113732870 (108M) (unauthoritative)\n",
      "\n",
      "schistosoma_haemato 100%[===================>] 108.46M  2.01MB/s    in 58s     \n",
      "\n",
      "2021-02-12 16:37:36 (1.89 MB/s) - ‘data/genome/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa.gz’ saved [113732870]\n",
      "\n",
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) [bwa_index] Pack FASTA... 6.36 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=742788110, availableWord=64265196\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 99324382 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 189907502 characters processed.\n",
      "[BWTIncConstructFromPacked] 30 iterations done. 270413998 characters processed.\n",
      "[BWTIncConstructFromPacked] 40 iterations done. 341964382 characters processed.\n",
      "[BWTIncConstructFromPacked] 50 iterations done. 405554542 characters processed.\n",
      "[BWTIncConstructFromPacked] 60 iterations done. 462069646 characters processed.\n",
      "[BWTIncConstructFromPacked] 70 iterations done. 512296366 characters processed.\n",
      "[BWTIncConstructFromPacked] 80 iterations done. 556933950 characters processed.\n",
      "[BWTIncConstructFromPacked] 90 iterations done. 596603966 characters processed.\n",
      "[BWTIncConstructFromPacked] 100 iterations done. 631858734 characters processed.\n",
      "[BWTIncConstructFromPacked] 110 iterations done. 663189246 characters processed.\n",
      "[BWTIncConstructFromPacked] 120 iterations done. 691031870 characters processed.\n",
      "[BWTIncConstructFromPacked] 130 iterations done. 715774414 characters processed.\n",
      "[BWTIncConstructFromPacked] 140 iterations done. 737761630 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 143 iterations.\n",
      "[bwa_index] 479.59 seconds elapse.\n",
      "[bwa_index] Update BWT... 5.88 sec\n",
      "[bwa_index] Pack forward-only FASTA... 4.65 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 205.80 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index data/genome/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa\n",
      "[main] Real time: 788.694 sec; CPU: 702.482 sec\n",
      "(PZQ-R_field) (PZQ-R_field) Using GATK jar /master/fcheval/local/pckg/python/miniconda3/envs/PZQ-R_field/share/gatk4-4.1.8.1-0/gatk-package-4.1.8.1-local.jar\n",
      "Running:\n",
      "    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -jar /master/fcheval/local/pckg/python/miniconda3/envs/PZQ-R_field/share/gatk4-4.1.8.1-0/gatk-package-4.1.8.1-local.jar CreateSequenceDictionary -R data/genome/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa\n",
      "INFO\t2021-02-12 16:51:43\tCreateSequenceDictionary\tOutput dictionary will be written in data/genome/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.dict\n",
      "16:51:43.354 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/master/fcheval/local/pckg/python/miniconda3/envs/PZQ-R_field/share/gatk4-4.1.8.1-0/gatk-package-4.1.8.1-local.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Fri Feb 12 16:51:43 CST 2021] CreateSequenceDictionary --REFERENCE data/genome/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa --TRUNCATE_NAMES_AT_WHITESPACE true --NUM_SEQUENCES 2147483647 --VERBOSITY INFO --QUIET false --VALIDATION_STRINGENCY STRICT --COMPRESSION_LEVEL 2 --MAX_RECORDS_IN_RAM 500000 --CREATE_INDEX false --CREATE_MD5_FILE false --GA4GH_CLIENT_SECRETS client_secrets.json --help false --version false --showHidden false --USE_JDK_DEFLATER false --USE_JDK_INFLATER false\n",
      "Feb 12, 2021 4:51:45 PM shaded.cloud_nio.com.google.auth.oauth2.ComputeEngineCredentials runningOnComputeEngine\n",
      "INFO: Failed to detect whether we are running on Google Compute Engine.\n",
      "[Fri Feb 12 16:51:45 CST 2021] Executing as fcheval@compute-1-1552.local on Linux 3.10.0-693.5.2.el7.x86_64 amd64; OpenJDK 64-Bit Server VM 1.8.0_152-release-1056-b12; Deflater: Intel; Inflater: Intel; Provider GCS is available; Picard version: Version:4.1.8.1\n",
      "[Fri Feb 12 16:51:48 CST 2021] picard.sam.CreateSequenceDictionary done. Elapsed time: 0.09 minutes.\n",
      "Runtime.totalMemory()=336068608\n",
      "Tool returned:\n",
      "0\n",
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) --2021-02-12 16:51:49--  ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS15/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS15.annotations.gff3.gz\n",
      "           => ‘data/genome/schistosoma_haematobium.PRJNA78265.WBPS15.annotations.gff3.gz’\n",
      "Resolving ftp.ebi.ac.uk... 193.62.197.74\n",
      "Connecting to ftp.ebi.ac.uk|193.62.197.74|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/databases/wormbase/parasite/releases/WBPS15/species/schistosoma_haematobium/PRJNA78265 ... done.\n",
      "==> SIZE schistosoma_haematobium.PRJNA78265.WBPS15.annotations.gff3.gz ... 6804872\n",
      "==> PASV ... done.    ==> RETR schistosoma_haematobium.PRJNA78265.WBPS15.annotations.gff3.gz ... done.\n",
      "Length: 6804872 (6.5M) (unauthoritative)\n",
      "\n",
      "schistosoma_haemato 100%[===================>]   6.49M   789KB/s    in 11s     \n",
      "\n",
      "2021-02-12 16:52:02 (599 KB/s) - ‘data/genome/schistosoma_haematobium.PRJNA78265.WBPS15.annotations.gff3.gz’ saved [6804872]\n",
      "\n",
      "(PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Download and unzip data\n",
    "wget -P \"$gdir\" ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS15/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa.gz\n",
    "pigz -d \"$gdir/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa.gz\"\n",
    "\n",
    "# Preparing indices\n",
    "bwa index \"$gdir/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa\"\n",
    "samtools faidx \"$gdir/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa\"\n",
    "gatk CreateSequenceDictionary -R \"$gdir/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa\"\n",
    "\n",
    "# Download the S. mansoni genome annotation\n",
    "wget -P \"$gdir\" ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS15/species/schistosoma_haematobium/PRJNA78265/schistosoma_haematobium.PRJNA78265.WBPS15.annotations.gff3.gz\n",
    "pigz -d \"$gdir/schistosoma_haematobium.PRJNA78265.WBPS15.annotations.gff3.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Known variants\n",
    "\n",
    "List of variants from the SmLE x SmHR crosses that segregated in a Mendelian fashion in F1 ([Valentim *et al.* 2013](https://doi.org/10.1126/science.1243106)). This list was generated from alignments using the v5 genome. This has been lifted over to the latest genome using the [flo pipeline](https://github.com/wurmlab/flo/tree/727f10b2b1c57a0514835d302d7f6345d3a34ffb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-11T04:04:26.444016Z",
     "start_time": "2021-01-11T04:04:20.050912Z"
    }
   },
   "outputs": [],
   "source": [
    "# Index file\n",
    "gatk IndexFeatureFile -I \"$gdir/sm_dbSNP_v7.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-14T19:41:52.029609Z",
     "start_time": "2021-01-14T19:41:51.708951Z"
    }
   },
   "outputs": [],
   "source": [
    "# For Freebayes\n",
    "bgzip -c \"$gdir/sm_dbSNP_v7.vcf\" > \"$gdir/sm_dbSNP_v7.vcf.gz\"\n",
    "tabix \"$gdir/sm_dbSNP_v7.vcf.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population files\n",
    "\n",
    "These populations files help to defined populations during variant calling with FreeBayes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-20T05:16:23.487Z"
    }
   },
   "outputs": [],
   "source": [
    "find \"$ldir_sm\" -type d -name \"*SAM*\"   | sed \"s|.*/||g\" | sed \"s|$|\\tUganda|g\" > data/Sm_pop\n",
    "find \"$ldir_sm\" -type d -name \"*Sm.TZ*\" | sed \"s|.*/||g\" | sed \"s|$|\\tTanzania|g\" >> data/Sm_pop\n",
    "find \"$ldir_sm\" -type d -name \"*Sm.SN*\" | sed \"s|.*/||g\" | sed \"s|$|\\tSenegal|g\" >> data/Sm_pop\n",
    "find \"$ldir_sm\" -type d -name \"*Sm.NE*\" | sed \"s|.*/||g\" | sed \"s|$|\\tNiger|g\" >> data/Sm_pop\n",
    "find \"$ldir_sm\" -type d -name \"*Sm.BR*\" | sed \"s|.*/||g\" | sed \"s|$|\\tBrazil|g\" >> data/Sm_pop\n",
    "find \"$ldir_sm\" -type d -name \"*Sm.OM*\" | sed \"s|.*/||g\" | sed \"s|$|\\tOman|g\" >> data/Sm_pop\n",
    "#find \"$ldir_sm\" -type d -name *Sr* | sed \"s|.*/||g\" | sed \"s|$|\\tSro|g\" >> data/Sm_pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-15T21:09:48.405539Z",
     "start_time": "2021-02-15T21:09:47.937251Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "find \"$ldir_sh\" -type d -name \"*Sh.TZ*\" | sed \"s|.*/||g\" | sed \"s|$|\\tTanzania|g\" > data/Sh_pop\n",
    "find \"$ldir_sh\" -type d -name \"*Sh.NE*\" | sed \"s|.*/||g\" | sed \"s|$|\\tNiger|g\" >> data/Sh_pop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene GFF files\n",
    "\n",
    "GFF file gene specific is required for some downstream analysis. This will be generated for *S. mansoni* and *S. haematobium* $TRP_{PZQ}$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-20T04:39:09.414147Z",
     "start_time": "2021-01-20T04:39:08.235548Z"
    }
   },
   "outputs": [],
   "source": [
    "# Gene code\n",
    "mygene=Smp_246790\n",
    "grep \"$mygene\" \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3\" > \"$gdir/$mygene.gff\"\n",
    "\n",
    "# Generate isoform specific gff\n",
    "myiso=$(grep -o \"$mygene\\.[0-9]*\" \"$gdir/$mygene.gff\" | sort | uniq)\n",
    "\n",
    "for i in $myiso\n",
    "do\n",
    "    awk '$3 == \"gene\"' \"$gdir/$mygene.gff\" > \"$gdir/$i.gff\"\n",
    "    grep -w \"$i\" \"$gdir/$mygene.gff\" | grep -v \"exon\" >> \"$gdir/$i.gff\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-15T21:29:10.595441Z",
     "start_time": "2021-02-15T21:29:04.844593Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Gene code\n",
    "mysh_gene=MS3_0012599\n",
    "grep \"$mysh_gene\" \"$gdir/schistosoma_haematobium.PRJNA78265.WBPS15.annotations.gff3\" > \"$gdir/$mysh_gene.gff\"\n",
    "\n",
    "# Generate isoform specific gff\n",
    "myiso=$(grep -o \"$mysh_gene\\.[0-9]*\" \"$gdir/$mysh_gene.gff\" | sort | uniq)\n",
    "\n",
    "for i in $myiso\n",
    "do\n",
    "    awk '$3 == \"gene\"' \"$gdir/$mysh_gene.gff\" > \"$gdir/$i.gff\"\n",
    "    grep -w \"$i\" \"$gdir/$mysh_gene.gff\" | grep -v \"exon\" >> \"$gdir/$i.gff\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequencing data processing\n",
    "\n",
    "### Snakemake pipeline\n",
    "\n",
    "The snakemake pipeline performs the following steps:\n",
    "* alignment of each library against the reference genome\n",
    "* marking duplicates\n",
    "* base quality recalibration (for *S mansoni* only)\n",
    "* generating alignment statistics\n",
    "* outputing $TRP_{PZQ}$ gene read depth for each library (*S. mansoni*: Smp_246790; *S. haematobium*: MS3_0012599)\n",
    "* outputing Z chromosome read depth for each library (for *S mansoni* only)\n",
    "* determining sex of each individual (for *S mansoni* only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-25T15:26:43.572362Z",
     "start_time": "2021-05-25T15:26:43.360639Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Status directory\n",
    "statdir=status\n",
    "[[ ! -d \"$statdir\" ]] && mkdir \"$statdir\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### *S. mansoni*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-20T03:57:03.036418Z",
     "start_time": "2021-01-20T03:57:02.821552Z"
    }
   },
   "outputs": [],
   "source": [
    "snakemake --snakefile Sm.smk --cluster \"qsub -V -cwd -o $statdir -j y -r y -pe smp 10 -S /bin/bash\" --jobs 1000 -w 120"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### *S. haematobium*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-20T03:57:03.036418Z",
     "start_time": "2021-01-20T03:57:02.821552Z"
    }
   },
   "outputs": [],
   "source": [
    "snakemake --snakefile Sh.smk --cluster \"qsub -V -cwd -o $statdir -j y -r y -pe smp 10 -S /bin/bash\" --jobs 1000 -w 120"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant calling\n",
    "\n",
    "We call the variants within the gene region across all samples directly from the BAM files using FreeBayes. This method is preferred because it can perform population defined calling."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-25T15:26:47.342443Z",
     "start_time": "2021-05-25T15:26:47.104661Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Output directory\n",
    "cdir=\"data/variant_calling\"\n",
    "[[ ! -d \"$cdir\" ]] && mkdir -p \"$cdir\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### *S. mansoni*\n",
    "\n",
    "##### Permissive calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T05:06:16.102298Z",
     "start_time": "2021-01-21T04:56:00.555221Z"
    }
   },
   "outputs": [],
   "source": [
    "# Variables\n",
    "mygenome=\"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "#mypos=SM_V7_3:600000-1000000\n",
    "mypos=SM_V7_3:630000-800000\n",
    "\n",
    "myfiles=$(find \"$ldir_sm\" -name *SAM*.bam -o -name *Sm*.bam | sort | tr \"\\n\" \" \") ## !! TO CHANGE SAMPLES\n",
    "\n",
    "myfilename=\"$cdir/PZQ-R_field\"\n",
    "\n",
    "\n",
    "# Command used for calling variants\n",
    "freebayes -f \"$mygenome\" \\\n",
    "   -r \"$mypos\" \\\n",
    "   -b $myfiles \\\n",
    "   -=          \\\n",
    "   --population data/Sm_pop \\\n",
    "   -q 20       \\\n",
    "   -m 30       \\\n",
    "   -@ \"$gdir/sm_dbSNP_v7.vcf.gz\" > \"${myfilename}.vcf\"\n",
    "   \n",
    "#     -F 0.4      \\\n",
    "\n",
    "pigz \"${myfilename}\".vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Stringent calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T16:18:26.402463Z",
     "start_time": "2021-05-20T16:18:14.041967Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) \n",
      "(PZQ-R_field) "
     ]
    }
   ],
   "source": [
    "# Variables\n",
    "mygenome=\"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "#mypos=SM_V7_3:600000-1000000\n",
    "mypos=SM_V7_3:630000-800000\n",
    "\n",
    "myfiles=$(find \"$ldir\" -name *SAM*.bam -o -name *Sm*.bam | sort | tr \"\\n\" \" \") ## !! TO CHANGE SAMPLES\n",
    "\n",
    "myfilename=\"$cdir/PZQ-R_field_restrictive\"\n",
    "\n",
    "\n",
    "# Command used for calling variants\n",
    "freebayes -f \"$mygenome\" \\\n",
    "   -r \"$mypos\" \\\n",
    "   -b $myfiles \\\n",
    "   -=          \\\n",
    "   --population data/Sm_pop \\\n",
    "   -q 20       \\\n",
    "   -m 30       \\\n",
    "   -F 0.3      \\\n",
    "   --min-coverage 4 \\\n",
    "   -@ \"$gdir/sm_dbSNP_v7.vcf.gz\" > \"${myfilename}.vcf\"\n",
    "  \n",
    "\n",
    "\n",
    "pigz \"${myfilename}\".vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-25T15:35:41.102929Z",
     "start_time": "2021-05-25T15:28:05.619384Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) WARNING(freebayes): Could not set the region of the variants input file to SM_V7_3:630000..800000\n",
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Variables\n",
    "mygenome=\"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "#mypos=SM_V7_3:600000-1000000\n",
    "mypos=SM_V7_3:630000-800000\n",
    "\n",
    "myfiles=$(find \"$ldir\" -name *SAM*.bam -o -name *Sm*.bam | sort | tr \"\\n\" \" \") ## !! TO CHANGE SAMPLES\n",
    "\n",
    "myfilename=\"$cdir/PZQ-R_field_restrictive2\"\n",
    "\n",
    "\n",
    "# Command used for calling variants\n",
    "freebayes -f \"$mygenome\" \\\n",
    "   -r \"$mypos\" \\\n",
    "   -b $myfiles \\\n",
    "   -=          \\\n",
    "   --population data/Sm_pop \\\n",
    "   -q 20       \\\n",
    "   -m 30       \\\n",
    "   -F 0.3      \\\n",
    "   --min-coverage 10 \\\n",
    "   -@ \"$gdir/sm_dbSNP_v7.vcf.gz\" > \"${myfilename}.vcf\"\n",
    "  \n",
    "\n",
    "\n",
    "pigz \"${myfilename}\".vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-25T15:38:57.146589Z",
     "start_time": "2021-05-25T15:38:57.038322Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/genome\n",
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "echo $gdir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### *S. haematobium*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-16T03:41:21.693634Z",
     "start_time": "2021-02-16T03:36:57.345217Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Variables\n",
    "mygenome=\"$gdir/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa\"\n",
    "mypos=scaffold020018:10279500-10358000\n",
    "\n",
    "myfiles=$(find \"$ldir_sh\" -name \"*.bam\" | sort | tr \"\\n\" \" \")\n",
    "\n",
    "myfilename=\"$cdir/Sh_PZQ-R_field\"\n",
    "\n",
    "\n",
    "# Command used for calling variants\n",
    "freebayes -f \"$mygenome\" \\\n",
    "   -r \"$mypos\" \\\n",
    "   -b $myfiles \\\n",
    "   -=          \\\n",
    "   --population data/Sh_pop \\\n",
    "   -q 20       \\\n",
    "   -m 30 > \"${myfilename}.vcf\"\n",
    "\n",
    "pigz \"${myfilename}\".vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyzing $SmTRP_{PZQ}$\n",
    "\n",
    "We will analyze the previously called variants on the gene only. This analysis includes:\n",
    "* Filtering out genotypes with low read depth,\n",
    "* Removing invariable sites,\n",
    "* Normalizing coordinates,\n",
    "* Use a custom script to summarize variants and genotype from the VCF file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T05:16:58.645883Z",
     "start_time": "2021-01-21T05:13:33.504609Z"
    }
   },
   "outputs": [],
   "source": [
    "myfilename=\"$cdir/PZQ-R_field\"\n",
    "\n",
    "# Remove genotypes with low read depths\n",
    "vcftools --gzvcf \"${myfilename}.vcf.gz\" \\\n",
    "    --minDP 4 \\\n",
    "    --recode  \\\n",
    "    --recode-INFO-all \\\n",
    "    --out \"${myfilename}.flt-dp4\"\n",
    "\n",
    "# Rename file\n",
    "mv \"${myfilename}.flt-dp4.recode.vcf\" \"${myfilename}.flt-dp4.vcf\"\n",
    "\n",
    "# Clean the VCF of invariable ref variants\n",
    "vcf-subset -a -e \"${myfilename}.flt-dp4.vcf\" > \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "\n",
    "# # Normalize VCF\n",
    "# vt normalize -o \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "#     -r \"$mygenome\" \\\n",
    "#     \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "## This fails with a core-dump\n",
    "\n",
    "## Maybe look at bcftools norm (with -f option)   !!! Install bcftools\n",
    "bcftools norm -o \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "    -f \"$mygenome\" \\\n",
    "    \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "\n",
    "##Another method of normalization\n",
    "# vcfleftalign -r \"$mygenome\" \\\n",
    "#     \"${myfilename}.flt-dp4.cleaned.vcf\" > \"${myfilename}.flt-dp4.cleaned.norm.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-08T19:27:47.775284Z",
     "start_time": "2021-06-08T19:27:47.564213Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Result folder\n",
    "rdir=\"results/1-reports\"\n",
    "[[ ! -d \"$rdir\" ]] && mkdir -p \"$rdir\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-21T05:20:42.965528Z",
     "start_time": "2021-01-21T05:20:41.291224Z"
    }
   },
   "outputs": [],
   "source": [
    "# Gene and isoforms\n",
    "mygene=Smp_246790\n",
    "myiso=$(grep -o \"$mygene\\.[0-9]*\" \"$gdir/$mygene.gff\" | sort | uniq)\n",
    "\n",
    "# Generate reports per isoform\n",
    "for i in $myiso\n",
    "do\n",
    "    qsub -V -cwd -o \"$statdir\" -j y -b y scripts/func-table-mut.sh -v \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "        -r \"$mygenome\" \\\n",
    "        -g \"$gdir/$i.gff\" \\\n",
    "        -p data/Sm_pop \\\n",
    "        -o \"$rdir/$i.flt.norm.tsv\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Stringent calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2021-05-25T16:07:42.522Z"
    }
   },
   "outputs": [],
   "source": [
    "myfilename=\"$cdir/PZQ-R_field_restrictive2\"\n",
    "\n",
    "# Remove genotypes with low read depths\n",
    "vcftools --gzvcf \"${myfilename}.vcf.gz\" \\\n",
    "    --minDP 4 \\\n",
    "    --recode  \\\n",
    "    --recode-INFO-all \\\n",
    "    --out \"${myfilename}.flt-dp4\"\n",
    "\n",
    "# Rename file\n",
    "mv \"${myfilename}.flt-dp4.recode.vcf\" \"${myfilename}.flt-dp4.vcf\"\n",
    "\n",
    "# Clean the VCF of invariable ref variants\n",
    "vcf-subset -a -e \"${myfilename}.flt-dp4.vcf\" > \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "\n",
    "# # Normalize VCF\n",
    "# vt normalize -o \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "#     -r \"$mygenome\" \\\n",
    "#     \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "## This fails with a core-dump\n",
    "\n",
    "## Maybe look at bcftools norm (with -f option)   !!! Install bcftools\n",
    "bcftools norm -o \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "    -f \"$mygenome\" \\\n",
    "    \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "\n",
    "##Another method of normalization\n",
    "# vcfleftalign -r \"$mygenome\" \\\n",
    "#     \"${myfilename}.flt-dp4.cleaned.vcf\" > \"${myfilename}.flt-dp4.cleaned.norm.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-25T16:10:50.278404Z",
     "start_time": "2021-05-25T16:10:48.767416Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) Your job 121955 (\"func-table-mut.sh\") has been submitted\n",
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) Your job 121956 (\"func-table-mut.sh\") has been submitted\n",
      "Your job 121957 (\"func-table-mut.sh\") has been submitted\n",
      "Your job 121958 (\"func-table-mut.sh\") has been submitted\n",
      "Your job 121959 (\"func-table-mut.sh\") has been submitted\n",
      "Your job 121960 (\"func-table-mut.sh\") has been submitted\n",
      "Your job 121961 (\"func-table-mut.sh\") has been submitted\n",
      "Your job 121962 (\"func-table-mut.sh\") has been submitted\n",
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "myfilename=\"$cdir/PZQ-R_field_restrictive2\"\n",
    "\n",
    "# Gene and isoforms\n",
    "mygene=Smp_246790\n",
    "myiso=$(grep -o \"$mygene\\.[0-9]*\" \"$gdir/$mygene.gff\" | sort | uniq)\n",
    "\n",
    "\n",
    "qsub -V -cwd -o \"$statdir\" -j y -b y scripts/func-table-mut.sh -v \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "        -r \"$mygenome\" \\\n",
    "        -g \"$gdir/$mygene.gff\" \\\n",
    "        -p data/Sm_pop \\\n",
    "        -o \"$rdir/${myfilename##*/}_${mygene}.flt.norm.tsv\"\n",
    "\n",
    "# Generate reports per isoform\n",
    "for i in $myiso\n",
    "do\n",
    "    qsub -V -cwd -o \"$statdir\" -j y -b y scripts/func-table-mut.sh -v \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "        -r \"$mygenome\" \\\n",
    "        -g \"$gdir/$i.gff\" \\\n",
    "        -p data/Sm_pop \\\n",
    "        -o \"$rdir/${myfilename##*/}_${i}.flt.norm.tsv\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-06-08T22:26:55.335008Z",
     "start_time": "2021-06-08T22:19:03.849375Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) \u001b[32mInfo:\u001b[00m Identifying samples related to Tanzania, Senegal, Niger, Brazil, Oman population(s)\n",
      "\u001b[32mInfo:\u001b[00m Extraction information from GFF\n",
      "\u001b[32mInfo:\u001b[00m Generating reference sequence\n",
      "sed: couldn't write 11202 items to stdout: Broken pipe\n",
      "\u001b[32mProgress:\u001b[00m [========================================] 100%\n",
      "\u001b[33mWarning:\u001b[00m Complex events were present in the VCF file. If you need to simplify them, use vcfalleleprimitives (with -g) from vcflib. However, the allele frequency of these simplified events might be incorrect.\n",
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "myfilename=\"$cdir/PZQ-R_field_restrictive2\"\n",
    "\n",
    "# Gene and isoforms\n",
    "mygene=Smp_246790\n",
    "myiso=$(grep -o \"$mygene\\.[0-9]*\" \"$gdir/$mygene.gff\" | sort | uniq)\n",
    "\n",
    "# Generate reports per isoform\n",
    "#for i in $myiso\n",
    "for i in ${mygene}.5\n",
    "do\n",
    "    scripts/func-table-mut.sh -v \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "        -r \"$mygenome\" \\\n",
    "        -g \"$gdir/$i.gff\" \\\n",
    "        -p data/Sm_pop \\\n",
    "        -a pop_cln \\\n",
    "        -o \"$rdir/${myfilename##*/}_${i}.flt.norm.tsv\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global coverage of the gene\n",
    "\n",
    "We inspect how well the gene and particularly the exons are covered. This is done by concatenating individual coverage of the gene and then plotting total read depth at each position. The concatenation step can take ~20 minutes or more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-12T20:56:18.906895Z",
     "start_time": "2021-02-12T20:39:36.708491Z"
    }
   },
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir2=\"results/2-Coverage\"\n",
    "[[ ! -d \"$rdir2\" ]] && mkdir -p \"$rdir2\"\n",
    "\n",
    "# Output file\n",
    "fl_out=\"$rdir2/SmTRP-PZQ.cov\"\n",
    "\n",
    "# List files\n",
    "fl_ls=($(find \"$ldir\" -type f -name \"*PZQ.cov\" | sort))\n",
    "\n",
    "# Create header\n",
    "echo -e \"#CHROM\\t$(cut -f1 \"${fl_ls[0]}\" | tr \"\\n\" \"\\t\" | sed \"s/\\t$//\")\" > \"$fl_out\"\n",
    "echo -e \"POS\\t$(cut -f2 \"${fl_ls[0]}\"| tr \"\\n\" \"\\t\" | sed \"s/\\t$//\")\" >> \"$fl_out\"\n",
    "\n",
    "for ((i = 0 ; i < ${#fl_ls[@]} ; i++))\n",
    "do\n",
    "    echo $i\n",
    "    echo -e \"$(dirname \"${fl_ls[$i]}\" | sed \"s,.*/,,\")\\t$(cut -f 3 \"${fl_ls[$i]}\" | tr \"\\n\" \"\\t\" | sed \"s/\\t$//\")\"  >> \"$fl_out\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting data\n",
    "\n",
    "Rscript scripts/SmTRP-PZQ_coverage.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutations of interest\n",
    "\n",
    "Some of the mutations are of high interest because located in critical sections of the protein (PZQ binding pocket, channel, TRP box). We list the samples that carry them in order to look at the reads within these samples to assess the likelihood of these mutations. This will also help identify samples that will require further investigation like PCR validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-01-25T17:55:43.118997Z",
     "start_time": "2021-01-25T17:54:48.030156Z"
    }
   },
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir3=\"results/3-mutations of interest\"\n",
    "[[ ! -d \"$rdir3\" ]] && mkdir -p \"$rdir3\"\n",
    "\n",
    "# mutations=\"p.T1394I p.Q1432H p.S1448Y p.G1458C p.H1522Q p.I1523N p.P1527T p.P1527Q p.Q1598K p.D1602Y p.D1606Y p.Q1609L p.Q1673K p.M1674I p.D1677Y p.H1680N p.P1683Q p.P1683R p.L1684I p.P1686T p.P1686Q p.P1687Q p.W1692L p.E1696* p.A1700D p.Q1704K\"\n",
    "mutations=\"p.H1522Q p.I1523N p.P1527T p.P1527Q p.Q1598K p.D1602Y p.D1606Y p.Q1609L p.Q1673K p.M1674I p.D1677Y p.H1680N p.P1686T p.P1686Q p.P1687Q\"\n",
    "\n",
    "# Add stop codon\n",
    "mutations+=\" $(awk '$8 ~ /*/ {print $8}' $rdir/Smp_246790.5.flt.norm.tsv)\"\n",
    "\n",
    "# Isoform of interest\n",
    "myiso=Smp_246790.5\n",
    "\n",
    "unset mylist\n",
    "for m in $mutations\n",
    "do\n",
    "echo $m\n",
    "    mypos=$(awk -v m=$m '$8 == m {print $1}' \"$rdir/$myiso.flt.norm.tsv\")\n",
    "    myspl=$(awk -v mypos=$mypos '$2 == mypos' \"${myfilename}.flt-dp4.cleaned.norm.vcf\"| tr \"\\t\" \"\\n\" | grep -n 0/1 | cut -d \":\" -f 1 | tr \"\\n\" \",\" | sed \"s/,$//\")\n",
    "    myspl=$(grep -m 1 \"#C\" \"${myfilename}.flt-dp4.cleaned.norm.vcf\" | cut -f $myspl)\n",
    "    mylist+=\"$m\\t$mypos\\t$myspl\"\"\\n\" \n",
    "done\n",
    "\n",
    "echo -ne \"$mylist\" > \"$rdir3/samples_list.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output read depth for each sample carrying mutations of interest\n",
    "Rscript scripts/variant_read_depth.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plots of the read depth for each mutation and sample are available in the folder labeled graphs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Primer design\n",
    "\n",
    "Interesting mutations from the calling data must be checked by Sanger sequencing. These mutations are located in specific exons. To sequence the full exons, we extract exon sequences with an additional 200 bp on each end, where primers should be designed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-09T20:04:08.658360Z",
     "start_time": "2021-02-09T20:04:02.081894Z"
    }
   },
   "outputs": [],
   "source": [
    "fl_out=\"$rdir3/exons_of_interest\"\n",
    "\n",
    "myexons=(3 4 12 23 25 27 29 34)\n",
    "myiso=\"Smp_246790.5\"\n",
    "\n",
    "grep \"$myiso\" \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3\" | \\\n",
    "    awk '$3 == \"CDS\"'  | \\\n",
    "    sed -n \"$(sed \"s/ /p;/g\" <<< ${myexons[@]})p\" | \\\n",
    "    awk '{print $1\"\\t\"$4-201\"\\t\"$5+200}' > \"$fl_out.bed\"\n",
    "\n",
    "bedtools getfasta -fi \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\" \\\n",
    "    -bed \"$fl_out.bed\" \\\n",
    "    -fo \"$fl_out.fas\"\n",
    "\n",
    "# Reformat fasta entry\n",
    "myentries=($(grep -n \">\" \"$fl_out.fas\" | cut -d \":\" -f 1))\n",
    "for ((i = 0 ; i < \"${#myexons[@]}\" ; i++))\n",
    "do\n",
    "    sed -i \"${myentries[$i]}s/>/>E${myexons[$i]} /\" \"$fl_out.fas\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sanger sequencing\n",
    "\n",
    "Sanger sequencing traces are analyzed using Consed and PolyPhred: The first align the data while the second perform the calling. PolyPhred output is then processed with a custom script in order to summarize the calling per sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data folder\n",
    "mkdir data/sanger_sequencing\n",
    "\n",
    "# Download the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:11:18.003687Z",
     "start_time": "2021-02-07T06:11:15.072648Z"
    }
   },
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir3b=\"$rdir3/sanger_sequencing\"\n",
    "[[ ! -d \"$rdir3b\" ]] && mkdir -p \"$rdir3b\"\n",
    "\n",
    "# Phred folders\n",
    "mkdir -p \"$rdir3b/data/\"{chromat_dir,edit_dir,phd_dir,poly_dir}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:11:18.003687Z",
     "start_time": "2021-02-07T06:11:15.072648Z"
    }
   },
   "outputs": [],
   "source": [
    "mygene=\"Smp_246790\"\n",
    "\n",
    "# Generate reference sequence\n",
    "mkdir \"$rdir3b/0-Ref_seq\"\n",
    "bedtools getfasta -fi \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\" \\\n",
    "    -bed <(grep \"$mygene\" \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3\" | awk '$3 == \"gene\"' | awk '{print $1\"\\t\"$4-1\"\\t\"$5}') \\\n",
    "    -fo \"$rdir3b/Smp_246790.fas\"\n",
    "\n",
    "# Rename fasta entry\n",
    "sed -i \"s/>.*/>$mygene/\" \"$rdir3/Smp_246790.fas\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-07T06:15:44.258775Z",
     "start_time": "2021-02-07T06:15:38.876037Z"
    }
   },
   "outputs": [],
   "source": [
    "pwd_old=$PWD\n",
    "\n",
    "# Generate ref sequence for phred\n",
    "cd \"$rdir3b/\"\n",
    "sudophred \"$mygene.fas\" -r -abi\n",
    "\n",
    "# Move ref\n",
    "mv \"$mygene.REF\" \"data/chromat_dir/\"\n",
    "mv \"$mygene.REF.phd.1\" \"data/phd_dir/\"\n",
    "\n",
    "cd \"$pwd_old\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare data\n",
    "\n",
    "We create links to the original data but with modified file names in order to informed Polyphred about the paired samples. Samples will be identified on the first 8 characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T17:45:03.119384Z",
     "start_time": "2021-05-20T17:44:55.521789Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Generate modified sample names for polyphred\n",
    "j=1\n",
    "unset spl_code\n",
    "for i in $(tail -n +2 \"$rdir3b/barcode_list.tsv\" | cut -f 1 | sort | uniq)\n",
    "do\n",
    "    spl_code+=\"$i\\t$(sed \"s/^/$(printf \"%02d\" $j)_/\" <<< \"$i\")\\n\"\n",
    "    ((j++))\n",
    "done\n",
    "\n",
    "# Generate exon specific folder and link to data\n",
    "while read line \n",
    "do\n",
    "    # Sample name\n",
    "    myspl=$(cut -f 1 <<< \"$line\")\n",
    "    myspl=$(echo -ne \"$spl_code\" | grep -w \"$myspl\" | cut -f2)\n",
    "    \n",
    "    # Barcode\n",
    "    mycode=$(cut -f 3 <<< \"$line\")\n",
    "    [[ \"$mycode\" == NA ]] && continue\n",
    "    \n",
    "    # Link files\n",
    "    ln -s \"$PWD/data/sanger_sequencing/$mycode\"*.scf \"$rdir3b/data/chromat_dir/${myspl}_$mycode.scf\"\n",
    "\n",
    "done < <(tail -n +2 \"$rdir3b/barcode_list.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-09T21:46:40.756484Z",
     "start_time": "2021-02-09T21:46:40.238240Z"
    }
   },
   "source": [
    "Run Phred and Consed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T17:49:10.065495Z",
     "start_time": "2021-05-20T17:48:50.538619Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) bash: pushd: results/2-mutations of interest/sanger_sequencing/data/edit_dir: No such file or directory\n",
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field)   ../chromat_dir/01_Sm.BR_PdV.1079.1_BXJ481.scf\n",
      "  ../chromat_dir/01_Sm.BR_PdV.1079.1_BXJ483.scf\n",
      "  ../chromat_dir/02_Sm.BR_PdV.1371.1_BXJ485.scf\n",
      "  ../chromat_dir/02_Sm.BR_PdV.1371.1_BXJ487.scf\n",
      "  ../chromat_dir/03_Sm.BR_PdV.1475_rep_BXJ489.scf\n",
      "  ../chromat_dir/03_Sm.BR_PdV.1475_rep_BXJ491.scf\n",
      "  ../chromat_dir/04_Sm.BR_PdV.2196.2_BXJ493.scf\n",
      "  ../chromat_dir/04_Sm.BR_PdV.2196.2_BXJ495.scf\n",
      "  ../chromat_dir/05_Sm.BR_PdV.2225.1_BXJ496.scf\n",
      "  ../chromat_dir/05_Sm.BR_PdV.2225.1_BXJ497.scf\n",
      "  ../chromat_dir/06_Sm.BR_PdV.2300.1_BXJ498.scf\n",
      "  ../chromat_dir/06_Sm.BR_PdV.2300.1_BXJ499.scf\n",
      "  ../chromat_dir/07_Sm.NE_Di186.1_BWK472.scf\n",
      "  ../chromat_dir/07_Sm.NE_Di186.1_BWK473.scf\n",
      "  ../chromat_dir/08_Sm.NE_Di68.2_BXJ500.scf\n",
      "  ../chromat_dir/08_Sm.NE_Di68.2_BXJ501.scf\n",
      "  ../chromat_dir/09_Sm.OM_She-OD1_BWK488.scf\n",
      "  ../chromat_dir/09_Sm.OM_She-OD1_BWK489.scf\n",
      "  ../chromat_dir/10_Sm.OM_She-OD2_BXJ480.scf\n",
      "  ../chromat_dir/10_Sm.OM_She-OD2_BXJ482.scf\n",
      "  ../chromat_dir/11_Sm.OM_Sig1-ONc4_BXJ502.scf\n",
      "  ../chromat_dir/11_Sm.OM_Sig1-ONc4_BXJ503.scf\n",
      "  ../chromat_dir/12_Sm.OM_Sig2-ONc10_BXJ504.scf\n",
      "  ../chromat_dir/12_Sm.OM_Sig2-ONc10_BXJ505.scf\n",
      "  ../chromat_dir/13_Sm.OM_Tib-H8_BWK490.scf\n",
      "  ../chromat_dir/13_Sm.OM_Tib-H8_BWK491.scf\n",
      "  ../chromat_dir/14_Sm.OM_Tib-ON1_BWK486.scf\n",
      "  ../chromat_dir/14_Sm.OM_Tib-ON1_BWK487.scf\n",
      "  ../chromat_dir/15_Sm.OM_Tib-ON13_BXJ484.scf\n",
      "  ../chromat_dir/15_Sm.OM_Tib-ON13_BXJ486.scf\n",
      "  ../chromat_dir/16_Sm.OM_Tib-ON14_BXJ506.scf\n",
      "  ../chromat_dir/16_Sm.OM_Tib-ON14_BXJ507.scf\n",
      "  ../chromat_dir/17_Sm.OM_Tib-ON8_BXJ508.scf\n",
      "  ../chromat_dir/17_Sm.OM_Tib-ON8_BXJ509.scf\n",
      "  ../chromat_dir/18_Sm.OM_Tib-T24_BWK474.scf\n",
      "  ../chromat_dir/18_Sm.OM_Tib-T24_BWK475.scf\n",
      "  ../chromat_dir/19_Sm.OM_Tib-T25_BXJ510.scf\n",
      "  ../chromat_dir/19_Sm.OM_Tib-T25_BXJ511.scf\n",
      "  ../chromat_dir/20_Sm.OM_Tib-T3M_BXJ512.scf\n",
      "  ../chromat_dir/20_Sm.OM_Tib-T3M_BXJ513.scf\n",
      "  ../chromat_dir/21_Sm.OM_Tib-T5F_BXJ514.scf\n",
      "  ../chromat_dir/21_Sm.OM_Tib-T5F_BXJ515.scf\n",
      "  ../chromat_dir/22_Sm.OM_Tib-T8M_BXJ516.scf\n",
      "  ../chromat_dir/22_Sm.OM_Tib-T8M_BXJ517.scf\n",
      "  ../chromat_dir/23_Sm.OM_WD1-ON19_BXJ518.scf\n",
      "  ../chromat_dir/23_Sm.OM_WD1-ON19_BXJ519.scf\n",
      "  ../chromat_dir/24_Sm.OM_Zee-OD21_BXJ488.scf\n",
      "  ../chromat_dir/24_Sm.OM_Zee-OD21_BXJ490.scf\n",
      "  ../chromat_dir/25_Sm.OM_Zee-OD24_BXJ520.scf\n",
      "  ../chromat_dir/25_Sm.OM_Zee-OD24_BXJ521.scf\n",
      "  ../chromat_dir/26_Sm.OM_Zee-OD25_BXJ522.scf\n",
      "  ../chromat_dir/26_Sm.OM_Zee-OD25_BXJ523.scf\n",
      "  ../chromat_dir/27_Sm.SN_Nd103.1_BWK482.scf\n",
      "  ../chromat_dir/27_Sm.SN_Nd103.1_BWK483.scf\n",
      "  ../chromat_dir/28_Sm.SN_Nd77.1_BXJ524.scf\n",
      "  ../chromat_dir/28_Sm.SN_Nd77.1_BXJ525.scf\n",
      "  ../chromat_dir/29_Sm.SN_Nd79.1_BWK484.scf\n",
      "  ../chromat_dir/29_Sm.SN_Nd79.1_BWK485.scf\n",
      "  ../chromat_dir/30_Sm.SN_Nd90.2s.1_BWK476.scf\n",
      "  ../chromat_dir/30_Sm.SN_Nd90.2s.1_BWK477.scf\n",
      "  ../chromat_dir/31_Sm.SN_Te55.1_BWK478.scf\n",
      "  ../chromat_dir/31_Sm.SN_Te55.1_BWK479.scf\n",
      "  ../chromat_dir/32_Sm.TZ_009.2.2_BWK480.scf\n",
      "  ../chromat_dir/32_Sm.TZ_009.2.2_BWK481.scf\n",
      "  ../chromat_dir/32_Sm.TZ_009.2.2_BXJ526.scf\n",
      "  ../chromat_dir/32_Sm.TZ_009.2.2_BXJ527.scf\n",
      "  ../chromat_dir/33_Sm.TZ_074N.7.2_BXJ528.scf\n",
      "  ../chromat_dir/33_Sm.TZ_074N.7.2_BXJ529.scf\n",
      "  ../chromat_dir/34_Sm.TZ_077.2.1_BXJ530.scf\n",
      "  ../chromat_dir/34_Sm.TZ_077.2.1_BXJ531.scf\n",
      "  ../chromat_dir/35_Sm.TZ_086.1.1_BXJ492.scf\n",
      "  ../chromat_dir/35_Sm.TZ_086.1.1_BXJ494.scf\n",
      "  ../chromat_dir/36_Sm.TZ_086.6.1_BXJ532.scf\n",
      "  ../chromat_dir/36_Sm.TZ_086.6.1_BXJ533.scf\n",
      "  ../chromat_dir/37_Sm.TZ_134.5.1_BXJ534.scf\n",
      "  ../chromat_dir/37_Sm.TZ_134.5.1_BXJ535.scf\n",
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pwd_old=$PWD\n",
    "\n",
    "# Moving to the working folder\n",
    "cd \"$rdir3b/data/edit_dir\"\n",
    "\n",
    "# Generating phd and poly files\n",
    "phred ../chromat_dir/*.scf -pd ../phd_dir/ -dd ../poly_dir/\n",
    "\n",
    "# Preparing the reference sequence\n",
    "phd2Ace.perl \"$mygene.REF.phd.1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T17:49:30.601116Z",
     "start_time": "2021-05-20T17:49:20.464275Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) -addNewReads will be run.\n",
      "no ~/consedrc file so no user resources will be used--that's ok\n",
      "no consedrc file so no project-specific resources--that's ok\n",
      "couldn't open readOrder.txt--that's ok\n",
      "50% done.  1 reads read so far...\n",
      "Now setting quality values\n",
      "Number of individual phd files read: 1\n",
      "Total reads in assembly: 1\n",
      "Finished setting quality values in 1 seconds \n",
      "Consed is detecting that there are 0 forward universal primer reads, 0 reverse universal primer reads, 1 walking reads, 0 pcr end reads, 0 reads of unknown type, and 0 reads did not have their type set in the phd file.  Is this correct?  If not, consed will not pick templates correctly for primers and autofinish will not function correctly.  You probably did not correct customize determineReadTypes.perl to conform to your read naming convention.  I suggest you read README.txt that came with consed and also examine the notes in determineReadTypes.perl.  To help you in figuring out what is wrong, on the Main Consed Window, point to Info, hold down mouse button 1, and release on Show Info For Each Read\n",
      "Version 29.0 (150417)\n",
      "ace file: Smp_246790.REF.ace\n",
      "after sortContigTemplateArrays\n",
      "\n",
      "\n",
      "--------------------------------------------------------\n",
      "Now running determineReadTypes.perl...\n",
      "--------------------------------------------------------\n",
      "\n",
      "\n",
      "files examined: 1, files processed: 1\n",
      "files examined: 2, files processed: 2\n",
      "files examined: 3, files processed: 3\n",
      "files examined: 4, files processed: 4\n",
      "files examined: 5, files processed: 5\n",
      "files examined: 6, files processed: 6\n",
      "files examined: 7, files processed: 7\n",
      "files examined: 8, files processed: 8\n",
      "files examined: 9, files processed: 9\n",
      "\n",
      "/master/fcheval/local/pckg/python/miniconda3/envs/PZQ-R_field/bin/consed.d/bin/cross_match newReadsaddNewReadsAutomated.210520.124924.cross.fasta /master/fcheval/local/pckg/python/miniconda3/envs/PZQ-R_field/bin/consed.d/lib/screenLibs/vector.seq -minmatch 12 -penalty -2 -minscore 20 -screen -tags \n",
      "cross_match version 0.990329\n",
      "Reading parameters ... 1.008 Mbytes allocated -- total 1.008 Mbytes\n",
      "Run date:time  210520:124927\n",
      " Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 1.003 Mbytes in 2 blocks\n",
      "Reading query file into memory ... Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 0.716 Mbytes in 2 blocks\n",
      "Complementing ... Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 0.716 Mbytes in 2 blocks\n",
      "Allocating align_entries ... Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 0.614 Mbytes in 2 blocks\n",
      "Reading quality files ...\n",
      "\n",
      "NO QUALITY FILE newReadsaddNewReadsAutomated.210520.124924.cross.fasta.qual WAS FOUND. REMAINING INPUT QUALITIES SET TO 15. Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 0.614 Mbytes in 2 blocks\n",
      "5.040 Mbytes allocated -- total 6.048 Mbytes\n",
      "Finding words ...\n",
      " 0  1014689\n",
      " 10  32810\n",
      " 20  311\n",
      " 30  154\n",
      " 40  138\n",
      " 50  446\n",
      " 60  21\n",
      " 70  3\n",
      " 130  2\n",
      " 300  1\n",
      " 310  1\n",
      "min no. words: 0 at 13\n",
      "max no. words: 304 at 1048575\n",
      "avg no. words: 0.0 Done\n",
      "Total space allocated: 6.048 Mbytes; currently free: 0.998 Mbytes in 3 blocks\n",
      "181262 words; after pruning: 81736\n",
      "Sorting words ... Done\n",
      "\n",
      "12  39011\n",
      "Finding subject matches ... Done\n",
      "Total space allocated: 6.048 Mbytes; currently free: 1.000 Mbytes in 4 blocks\n",
      "# pass words: 0, #fail words: 0\n",
      "Quickalign: 0 successes, 0 failures\n",
      "0 SWAT alignments performed. 0 pairs have score >= 20\n",
      "Making reversed pairs ... Done\n",
      "Total space allocated: 6.048 Mbytes; currently free: 5.603 Mbytes in 3 blocks\n",
      "Printing matches ...  Done\n",
      "Total space allocated: 6.048 Mbytes; currently free: 5.603 Mbytes in 4 blocks\n",
      "Checking pairs ... Done\n",
      "Total space allocated: 6.048 Mbytes; currently free: 5.603 Mbytes in 3 blocks\n",
      "Total space allocated: 6.048 Mbytes; currently free: 5.603 Mbytes in 3 blocks\n",
      "Total # pairs: 0, size: 0.000 Mbytes\n",
      "Total # segment blocks: 0, size: 0.000 Mbytes\n",
      "Total # diffs: 0, in 0 lists, size: 0.000 Mbytes\n",
      "\n",
      "using ace file format 2 to construct contigs\n",
      "now running cross_match between the new reads and the existing contigs\n",
      "\n",
      "/master/fcheval/local/pckg/python/miniconda3/envs/PZQ-R_field/bin/consed.d/bin/cross_match newReadsaddNewReadsAutomated.210520.124924.cross.fasta.screen Smp_246790.REF.aceaddNewReadsAutomated.210520.124924.cross.contigs -discrep_lists -tags -masklevel 0 \n",
      "cross_match version 0.990329\n",
      "Reading parameters ... 1.008 Mbytes allocated -- total 1.008 Mbytes\n",
      "Run date:time  210520:124927\n",
      " Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 1.003 Mbytes in 2 blocks\n",
      "Reading query file into memory ... Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 0.716 Mbytes in 2 blocks\n",
      "Complementing ... Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 0.716 Mbytes in 2 blocks\n",
      "Allocating align_entries ... Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 0.614 Mbytes in 2 blocks\n",
      "Reading quality files ...\n",
      "\n",
      "NO QUALITY FILE newReadsaddNewReadsAutomated.210520.124924.cross.fasta.screen.qual WAS FOUND. REMAINING INPUT QUALITIES SET TO 15. Done\n",
      "Total space allocated: 1.008 Mbytes; currently free: 0.614 Mbytes in 2 blocks\n",
      "5.040 Mbytes allocated -- total 6.048 Mbytes\n",
      "Finding words ...\n",
      " 0  1015553\n",
      " 10  31958\n",
      " 20  303\n",
      " 30  155\n",
      " 40  136\n",
      " 50  447\n",
      " 60  19\n",
      " 70  1\n",
      " 100  2\n",
      " 270  2\n",
      "min no. words: 0 at 13\n",
      "max no. words: 265 at 1048575\n",
      "avg no. words: 0.0 Done\n",
      "Total space allocated: 6.048 Mbytes; currently free: 1.008 Mbytes in 3 blocks\n",
      "181262 words; after pruning: 79836\n",
      "Sorting words ... Done\n",
      "\n",
      "14  39825\n",
      "Finding subject matches ...\n",
      "seq_buffer expanded\n",
      "seq_buffer expanded1.008 Mbytes allocated -- total 7.056 Mbytes\n",
      "1.008 Mbytes allocated -- total 8.064 Mbytes\n",
      "1.008 Mbytes allocated -- total 9.072 Mbytes\n",
      "1.008 Mbytes allocated -- total 10.080 Mbytes\n",
      "1.008 Mbytes allocated -- total 11.088 Mbytes\n",
      "1.008 Mbytes allocated -- total 12.096 Mbytes\n",
      "1.008 Mbytes allocated -- total 13.104 Mbytes\n",
      "1.008 Mbytes allocated -- total 14.112 Mbytes\n",
      " Done\n",
      "Total space allocated: 14.112 Mbytes; currently free: 8.288 Mbytes in 17 blocks\n",
      "# pass words: 406, #fail words: 381\n",
      "Quickalign: 87 successes, 0 failures\n",
      "425 SWAT alignments performed. 87 pairs have score >= 30\n",
      "Making reversed pairs ... Done\n",
      "Total space allocated: 14.112 Mbytes; currently free: 12.878 Mbytes in 16 blocks\n",
      "Printing matches ...  Done\n",
      "Total space allocated: 14.112 Mbytes; currently free: 12.793 Mbytes in 17 blocks\n",
      "Checking pairs ... Done\n",
      "Total space allocated: 14.112 Mbytes; currently free: 12.794 Mbytes in 16 blocks\n",
      "Total space allocated: 14.112 Mbytes; currently free: 12.794 Mbytes in 16 blocks\n",
      "Total # pairs: 1000, size: 0.056 Mbytes\n",
      "Total # segment blocks: 1, size: 0.080 Mbytes\n",
      "Total # diffs: 1738, in 87 lists, size: 0.002 Mbytes\n",
      "\n",
      "done running cross_match\n",
      "starting insertPadsInContigs 6\n",
      "Inserting pads in contigs to accommodate insertions in new reads...Done\n",
      "ending insertPadsInContigs 0\n",
      "Inserting pads in reads and setting read bases...\n",
      "see Smp_246790.REF.210520.124930.nav\n",
      "writing Smp_246790.REF.ace.1\n",
      "Wrote new ace file: Smp_246790.REF.ace.1\n",
      "See log file: Smp_246790.REF.210520.124923.out\n",
      "(PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Aligning reads to the reference\n",
    "find ../chromat_dir/* -name \"*.scf\" -exec basename {} \\; > reads_to_add.fof\n",
    "consed -ace \"$mygene.REF.ace\" -addNewReads reads_to_add.fof -newAceFilename \"$mygene.REF.ace.1\"\n",
    "cd \"$pwd_old\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T17:52:36.298287Z",
     "start_time": "2021-05-20T17:52:36.077306Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) /data/infectious/schistosome/06 - PZQ resistance/2021-01-06 PZQ field variants\n",
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "cd \"/data/infectious/schistosome/06 - PZQ resistance/2021-01-06 PZQ field variants/\"\n",
    "echo $PWD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run PolyPhred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T17:53:34.929595Z",
     "start_time": "2021-05-20T17:52:54.945687Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) \n",
      "POLYPHRED  Version 6.18\n",
      "--------------------------------------------------------------\n",
      "Writing output to file /data/infectious/schistosome/06 - PZQ resistance/2021-01-06 PZQ field variants/results/2-mutations of interest/sanger_sequencing/data/edit_dir/Smp_246790.s70-q25.snp.polyphred.out\n",
      "\n",
      "Reading the ACE file /data/infectious/schistosome/06 - PZQ resistance/2021-01-06 PZQ field variants/results/2-mutations of interest/sanger_sequencing/data/edit_dir/Smp_246790.REF.ace.1\n",
      "Reading contig `Smp_246790.REF-Contig'\n",
      "Using sequence Smp_246790.REF as reference sequence.\n",
      "Processing the contig Smp_246790.REF-Contig\n",
      "\u001b[2K   0% done\u001b[A\n",
      "\u001b[2K   1% done (reading sequences)\u001b[A\n",
      "\u001b[2K   2% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   3% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   4% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   5% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   6% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   7% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   8% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   9% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  10% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  11% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  12% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  13% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  14% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  15% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  16% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  17% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  18% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  19% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  20% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  21% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  22% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  23% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  24% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  25% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  26% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  27% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  28% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  29% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  30% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  31% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  32% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  33% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  34% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  35% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  36% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  37% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  38% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  39% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  40% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  41% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  42% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  43% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  44% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  45% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  46% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  47% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  48% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  49% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  50% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  51% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  52% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  53% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  54% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  55% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  56% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  57% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  58% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  59% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  60% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  61% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  62% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  63% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  64% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  65% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  66% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  67% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  68% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  69% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  70% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  71% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  72% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  73% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  74% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  75% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  76% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  77% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  78% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  79% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  80% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  81% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  82% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  83% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  84% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  85% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  86% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  87% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  88% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  89% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  90% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  91% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  92% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  93% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  94% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  95% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  96% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  97% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  98% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  99% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K 100% done (searching for SNPs)\u001b[A\n",
      "\n",
      "\n",
      "Updated the PHD files.\n",
      "Updating the ACE file /data/infectious/schistosome/06 - PZQ resistance/2021-01-06 PZQ field variants/results/2-mutations of interest/sanger_sequencing/data/edit_dir/Smp_246790.REF.ace.1\n",
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) \n",
      "POLYPHRED  Version 6.18\n",
      "--------------------------------------------------------------\n",
      "Writing output to file /data/infectious/schistosome/06 - PZQ resistance/2021-01-06 PZQ field variants/results/2-mutations of interest/sanger_sequencing/data/edit_dir/Smp_246790.s70-q25.indel.polyphred.out\n",
      "\n",
      "Reading the ACE file /data/infectious/schistosome/06 - PZQ resistance/2021-01-06 PZQ field variants/results/2-mutations of interest/sanger_sequencing/data/edit_dir/Smp_246790.REF.ace.1\n",
      "Reading contig `Smp_246790.REF-Contig'\n",
      "Warning: skipping reference sequence Smp_246790.REF\n",
      "Processing the contig Smp_246790.REF-Contig\n",
      "Searching for indel sites with insertion/deletion lengths up to 30/30\n",
      "\u001b[2K   0% done\u001b[A\n",
      "\u001b[2K   1% done (reading sequences)\u001b[A\n",
      "\u001b[2K   2% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   3% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   4% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   5% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   6% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   7% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   8% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K   9% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  10% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  11% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  12% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  13% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  14% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  15% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  16% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  17% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  18% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  19% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  20% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  21% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  22% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  23% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  24% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  25% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  26% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  27% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  28% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  29% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  30% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  31% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  32% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  33% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  34% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  35% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  36% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  37% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  38% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  39% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  40% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  41% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  42% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  43% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  44% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  45% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  46% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  47% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  48% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  49% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  50% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  51% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  52% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  53% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  54% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  55% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  56% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  57% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  58% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  59% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  60% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  61% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  62% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  63% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  64% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  65% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  66% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  67% done (searching for SNPs)\u001b[A\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2K  68% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  69% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  70% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  71% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  72% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  73% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  74% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  75% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  76% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  77% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  78% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  79% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  80% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  81% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  82% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  83% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  84% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  85% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  86% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  87% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  88% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  89% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  90% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  91% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  92% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  93% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  94% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  95% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  96% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  97% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  98% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K  99% done (searching for SNPs)\u001b[A\n",
      "\u001b[2K 100% done (searching for SNPs)\u001b[A\n",
      "\n",
      "\n",
      "Updated the PHD files.\n",
      "Indel navigator file written to /data/infectious/schistosome/06 - PZQ resistance/2021-01-06 PZQ field variants/results/2-mutations of interest/sanger_sequencing/data/edit_dir/Smp_246790.s70-q25.indel.nav\n",
      "Updating the ACE file /data/infectious/schistosome/06 - PZQ resistance/2021-01-06 PZQ field variants/results/2-mutations of interest/sanger_sequencing/data/edit_dir/Smp_246790.REF.ace.1\n",
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "score_snp=70 #90\n",
    "score_indel=70 #80\n",
    "qual=25 #40\n",
    "\n",
    "# SNP scoring\n",
    "name=$mygene.s${score_snp}-q$qual\n",
    "polyphred -d \"$rdir3b/data\" -extended_genotype -n $name.snp.nav -o $name.snp.polyphred.out -q $qual -refcomp -s 1 8 -score $score_snp\n",
    "\n",
    "# Indel scoring\n",
    "name=$mygene.s${score_indel}-q$qual\n",
    "polyphred -d \"$rdir3b/data\" -extended_genotype -i -inav $name.indel.nav -o $name.indel.polyphred.out -q $qual -s 1 8 -iscore $score_indel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T17:54:54.228344Z",
     "start_time": "2021-05-20T17:54:34.491359Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:\u001b[00m [========================================] 100%\n",
      "\u001b[32mProgress:\u001b[00m [==========                              ] 25%\u001b[33mWarning:\u001b[00m Several genotype detected for sample A at site 128723. Skipping...\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample A at site 128806. Skipping...\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample A at site 128969. Skipping...\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample A at site 129073. Skipping...\n",
      "\u001b[32mProgress:\u001b[00m [====================                    ] 50%\u001b[33mWarning:\u001b[00m Several genotype detected for sample C at site 128779. Skipping...\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample C at site 129073. Skipping...\n",
      "\u001b[32mProgress:\u001b[00m [==============================          ] 75%\u001b[33mWarning:\u001b[00m Several genotype detected for sample G at site 128806. Skipping...\n",
      "\u001b[32mProgress:\u001b[00m [========================================] 100%\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample G at site 128806. Skipping...\n",
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Genotype extractions\n",
    "scripts/trim_polyphred.sh -i \"$rdir3b/data/edit_dir/$mygene.s${score_snp}-q$qual.snp.polyphred.out\" \\\n",
    "    -o \"$rdir3b/$mygene.SNP.gt.tsv\" \\\n",
    "    -p 2 3 \\\n",
    "    -d \"_\"\n",
    "\n",
    "scripts/trim_polyphred.sh -i \"$rdir3b/data/edit_dir/$mygene.s${score_indel}-q$qual.indel.polyphred.out\" \\\n",
    "    -o \"$rdir3b/$mygene.INDEL.gt.tsv\" \\\n",
    "    -p 2 3 \\\n",
    "    -d \"_\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-05-20T18:16:43.554917Z",
     "start_time": "2021-05-20T18:16:41.779853Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sed: -e expression #1, char 21: unknown command: `E'\n",
      "\u001b[32mProgress:\u001b[00m [==========                              ] 25%\u001b[33mWarning:\u001b[00m Several genotype detected for sample A at site 128723. Skipping...\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample A at site 128806. Skipping...\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample A at site 128969. Skipping...\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample A at site 129073. Skipping...\n",
      "\u001b[32mProgress:\u001b[00m [====================                    ] 50%\u001b[33mWarning:\u001b[00m Several genotype detected for sample C at site 128779. Skipping...\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample C at site 129073. Skipping...\n",
      "\u001b[32mProgress:\u001b[00m [==============================          ] 75%\u001b[33mWarning:\u001b[00m Several genotype detected for sample G at site 128806. Skipping...\n",
      "\u001b[32mProgress:\u001b[00m [========================================] 100%\n",
      "\u001b[33mWarning:\u001b[00m Several genotype detected for sample G at site 128806. Skipping...\n",
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "scripts/trim_polyphred.sh -i \"$rdir3b/data/edit_dir/$mygene.s${score_indel}-q$qual.indel.polyphred.out\" \\\n",
    "    -o \"$rdir3b/$mygene.INDEL.gt.tsv\" \\\n",
    "    -p 2 3 \\\n",
    "    -d \"_\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyzing $ShTRP_{PZQ}$\n",
    "\n",
    "We will analyze the previously called variants on the gene only. This analysis includes:\n",
    "* Filtering out genotypes with low read depth,\n",
    "* Removing invariable sites,\n",
    "* Normalizing coordinates,\n",
    "* Use a custom script to summarize variants and genotype from the VCF file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-16T05:09:40.435595Z",
     "start_time": "2021-02-16T05:09:26.054601Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) \n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf data/variant_calling/Sh_PZQ-R_field.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--minDP 4\n",
      "\t--out data/variant_calling/Sh_PZQ-R_field.flt-dp4\n",
      "\t--recode\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 96 out of 96 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 1913 out of a possible 1913 Sites\n",
      "Run Time = 2.00 seconds\n",
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) Lines total/modified/skipped:\t780/57/0\n",
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "myfilename=\"$cdir/Sh_PZQ-R_field\"\n",
    "\n",
    "# Remove genotypes with low read depths\n",
    "vcftools --gzvcf \"${myfilename}.vcf.gz\" \\\n",
    "    --minDP 4 \\\n",
    "    --recode  \\\n",
    "    --recode-INFO-all \\\n",
    "    --out \"${myfilename}.flt-dp4\"\n",
    "\n",
    "# Rename file\n",
    "mv \"${myfilename}.flt-dp4.recode.vcf\" \"${myfilename}.flt-dp4.vcf\"\n",
    "\n",
    "# Clean the VCF of invariable ref variants\n",
    "vcf-subset -a -e \"${myfilename}.flt-dp4.vcf\" > \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "\n",
    "# Maybe look at bcftools norm (with -f option)   !!! Install bcftools\n",
    "bcftools norm -o \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "    -f \"$mygenome\" \\\n",
    "    \"${myfilename}.flt-dp4.cleaned.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-16T05:13:56.264129Z",
     "start_time": "2021-02-16T05:13:55.405531Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) Your job 33810 (\"func-table-mut.sh\") has been submitted\n",
      "(PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Gene and isoforms\n",
    "mysh_gene=MS3_0012599\n",
    "myiso=$(grep -o \"$mysh_gene\\.[0-9]*\" \"$gdir/$mysh_gene.gff\" | sort | uniq)\n",
    "\n",
    "\n",
    "# Generate reports per isoform\n",
    "for i in $myiso\n",
    "do\n",
    "    qsub -V -cwd -o \"$statdir\" -j y -b y scripts/func-table-mut.sh -v \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "        -r \"$mygenome\" \\\n",
    "        -g \"$gdir/$i.gff\" \\\n",
    "        -p data/Sh_pop \\\n",
    "        -o \"$rdir/$i.flt.norm.tsv\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-16T05:33:54.088930Z",
     "start_time": "2021-02-16T05:33:53.761020Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Result folder\n",
    "rdir3=\"results/3-mutations of interest\"\n",
    "[[ ! -d \"$rdir3\" ]] && mkdir -p \"$rdir3\"\n",
    "\n",
    "fl_out=\"$rdir3/Sh_exons_of_interest\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-02-16T05:43:35.134864Z",
     "start_time": "2021-02-16T05:43:34.499647Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) (PZQ-R_field) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "grep \"$myiso\" \"$gdir/schistosoma_haematobium.PRJNA78265.WBPS15.annotations.gff3\" | \\\n",
    "    awk '$3 == \"CDS\"'  > \"$fl_out.bed\"\n",
    "#     sed -n \"$(sed \"s/ /p;/g\" <<< ${myexons[@]})p\" | \\\n",
    "#     awk '{print $1\"\\t\"$4-201\"\\t\"$5+200}' > \"$fl_out.bed\"\n",
    "\n",
    "bedtools getfasta -fi \"$gdir/schistosoma_haematobium.PRJNA78265.WBPS15.genomic.fa\" \\\n",
    "    -bed \"$fl_out.bed\" \\\n",
    "    -fo \"$fl_out.fas\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "384px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
