{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of natural variations in *Sm.TRPM$_{PZQ}$*\n",
    "\n",
    "\n",
    "## Aim\n",
    "\n",
    "Our genome-wide association study revealed a quantitative trait locus (QTL) on chromosome 3 of *S. mansoni* linked to praziquantel (PZQ) response. This region contains the gene *Sm.TRPM$_{PZQ}$* (Smp_246790) which is likely the target of PZQ. We score and investigate variants of this gene from natural populations in order to have a comprehensive view of those that could impact PZQ efficacy. We used data from different locations in New and Old World:\n",
    "* Brazil (original from the associated publication)\n",
    "* Senegal\n",
    "* Niger\n",
    "* Tanzania\n",
    "* Uganda (data generated by Sanger https://doi.org/10.3389/fgene.2019.00826)\n",
    "* Oman\n",
    "\n",
    "\n",
    "## Environment and data\n",
    "\n",
    "### Environment\n",
    "\n",
    "Creating a conda environment improves reproducibility by installing specific versions of the programs used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "conda env create -f .env/env.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cell below must be run each time a new Jupyter session is run or when the kernel is rebooted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Activate the environment\n",
    "source $(sed \"s,/bin/conda,,\" <<<$CONDA_EXE)/etc/profile.d/conda.sh\n",
    "conda activate PZQ-R_field\n",
    "\n",
    "# Remove potential variable interferences\n",
    "export PERL5LIB=\"\"\n",
    "export PYTHONNOUSERSITE=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cell below must be run only once at the time of the environment creation. The R script install the R dependencies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Installing needed R packages\n",
    "Rscript \".env/R package dependencies.R\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Phred/Phrap/Consed and PolyPhred\n",
    "\n",
    "[Phred/Phrap/Consed](http://www.phrap.org/phredphrapconsed.html) and [PolyPhred](https://droog.gs.washington.edu/polyphred/) software allow analysis of Sanger sequencing traces by performing alignment and genotyping, respectively. They are freely available for academic used. However, this requires to contact the authors in order to get a copy ([here](http://www.phrap.org/consed/consed.html#howToGet) for Phred/Phrap/Consed, [here](https://droog.gs.washington.edu/polyphred/poly_get.html) for PolyPhred).\n",
    "\n",
    "Once the copies obtained, they must be placed in the download folder created with the command below. Then programs will be uncompressed and compiled using the cells below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder in which the software archives must be located\n",
    "ddir=\"download\"\n",
    "[[ ! -d \"$ddir\" ]] && mkdir -p \"$ddir\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phred\n",
    "\n",
    "Below is the procedure to compile `phred` and `daev`. This requires to have access to a terminal to execute some of the commands below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myphred=\"phred-dist-020425.c-acd.tar.Z\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/phred\" ]] && mkdir -p \"$ddir/phred\"\n",
    "tar -C \"$ddir/phred\" -xzvf \"$ddir/$myphred\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "\n",
    "cd \"$ddir/phred\"\n",
    "make\n",
    "make daev\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy the executables to the conda environment\n",
    "cp -a \"$ddir/phred/phred\" \"$ddir/phred/daev\" \"$CONDA_PREFIX/bin/\"\n",
    "\n",
    "# Copy the needed file\n",
    "cp -a \"$ddir/phred/phredpar.dat\" \"$CONDA_PREFIX/etc/phredpar.dat\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the machine and chemistry used by the sequencing facility to the phredpar file\n",
    "## Machine model and related information are available in the header of scf files\n",
    "sed -i '/^begin/,/^# Chromatograms/ s/^#$/\"KB_3730_POP7_BDTv3.mob\"        terminator      big-dye                 ABI_3700\\n#/' \"$CONDA_PREFIX/etc/phredpar.dat\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phd2fasta\n",
    "\n",
    "Below is the procedure to compile `phd2fasta`. This requires to have access to a terminal to execute some of the commands below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myphd=\"phd2fasta-acd-dist.130911.tar.gz\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/phd\" ]] && mkdir -p \"$ddir/phd\"\n",
    "tar -C \"$ddir/phd\" -xzvf \"$ddir/$myphd\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "\n",
    "cd \"$ddir/phd\"\n",
    "make\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy the executable to the conda environment\n",
    "cp -a \"$ddir/phd/phd2fasta\" \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Phrap\n",
    "\n",
    "Below is the procedure to compile `phrap`. This requires to have access to a terminal to execute some of the commands below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myphrap=\"distrib.tar.Z\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/phrap.d\" ]] && mkdir -p \"$ddir/phrap.d\"\n",
    "tar -C \"$ddir/phrap.d\" -xzvf \"$ddir/$myphrap\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "\n",
    "cd \"$ddir/phrap.d\"\n",
    "make\n",
    "make manyreads\n",
    "cd -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean and copy the folder to the conda environment\n",
    "rm \"$ddir/phrap.d/\"*.{c,h,o}\n",
    "cp -aR \"$ddir/phrap.d/\" \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ln -s \"$CONDA_PREFIX/bin/phrap.d/\"{cross_match,cross_match.manyreads,loco,phrap,phrap.longreads,phrap.manyreads,phrapview,swat} \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Consed\n",
    "\n",
    "Below is the procedure to compile `consed`. This requires to have access to a terminal to execute some of the commands below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mycs=\"consed_linux.tar.gz\"\n",
    "\n",
    "## Decompress archive\n",
    "[[ ! -d \"$ddir/consed\" ]] && mkdir -p \"$ddir/consed\"\n",
    "tar -C \"$ddir/consed\" -xzvf \"$ddir/$mycs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in \"$ddir/consed/\"consed*\n",
    "do\n",
    "    ./\"$i\" -v\n",
    "    [[ $? -eq 0 ]] && prog=$(basename \"$i\") && break\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct small bug in install script (when whitespace is present in current path)\n",
    "sed -i 's|cp $szDownloadDirectory|cp \\\\\"$szDownloadDirectory\\\\\"|g' \"$ddir/consed/installConsed.perl\"\n",
    "sed -i 's|cp $szNewDir|cp \\\\\"$szNewDir\\\\\"|g' \"$ddir/consed/installConsed.perl\""
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Run this cell from a terminal (compilation does not seem to work from the notebook)\n",
    "cd \"$ddir/consed/\"\n",
    "export CONSED_HOME=\"$CONDA_PREFIX/bin/consed.d\"\n",
    "mkdir \"$CONSED_HOME\"\n",
    "./installConsed.perl \"$prog\" \"$CONSED_HOME\"\n",
    "cd -\n",
    "ln -s \"$CONDA_PREFIX/bin/consed.d/bin/consed\" \"$CONDA_PREFIX/bin/consed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ln -s \"$CONDA_PREFIX/bin/phrap.d/\"{cross_match,cross_match.manyreads} \"$CONDA_PREFIX/bin/consed.d/bin\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ln -s \"$CONDA_PREFIX/bin/phred\" \"$CONDA_PREFIX/bin/consed.d/bin\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ln -s \"$CONDA_PREFIX/etc/phredpar.dat\" \"$CONDA_PREFIX/bin/consed.d/lib\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### PolyPhred\n",
    "\n",
    "Below is the procedure to compile `polyphred`. This requires to have access to a terminal to execute some of the commands below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mypp=\"polyphred-6.18-binary-x86_64-unknown-linux-gnu.tar.gz\"\n",
    "\n",
    "## Decompress archive\n",
    "tar -C \"$CONDA_PREFIX/bin/\" -xzvf \"$ddir/$mypp\"\n",
    "\n",
    "# Link executable\n",
    "ln -s \"$CONDA_PREFIX/bin/polyphred-6.18-binary-x86_64-unknown-linux-gnu/bin/\"* \"$CONDA_PREFIX/bin/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct perl location\n",
    "sed -i '1s|^#!.*|#!/usr/bin/env perl|' \"$CONDA_PREFIX/bin/polyphred-6.18-binary-x86_64-unknown-linux-gnu/bin/phredPhrap.pl\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Environment variable\n",
    "\n",
    "In order to have `phred` and `consed` working as expected, two environment variables must be exported each time a new Jupyter session is run or when the kernel is rebooted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "export PHRED_PARAMETER_FILE=\"$CONDA_PREFIX/etc/phredpar.dat\"\n",
    "export CONSED_HOME=\"$CONDA_PREFIX/bin/consed.d\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exome sequence data\n",
    "\n",
    "This step downloads the fastq files of the different samples from the SRA repository."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data directory\n",
    "ldir=\"data/libraries\"\n",
    "[[ ! -d \"$ldir\" ]] && mkdir -p \"$ldir\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Samples from Uganda\n",
    "\n",
    "Unfortunately, these samples were not informative regarding our gene of interest: only few reads mapped to the gene and no variants were scored confidently. Therefore downloading this data can be skipped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bioproject\n",
    "bioproject=ERP114942\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Keep S. mansoni samples only\n",
    "fds=$(head -1 runinfo | tr \",\" \"\\n\" | grep -w -n \"ScientificName\" | cut -d \":\" -f 1)\n",
    "awk -v fds=$fds -F \",\" '$fds == \"Schistosoma mansoni\"' runinfo > runinfo_sm\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir/$fln/\" ]] && mkdir -p \"$ldir/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: dowloading problem\" >> \"$ldir/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir/$fln/\" --split-files \"$ldir/$fln/$run\"\n",
    "    rm \"$ldir/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir/$fln/${run}_1.fastq\" \"$ldir/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir/$fln/${run}_2.fastq\" \"$ldir/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < runinfo_sm\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir/\"*/*\n",
    "\n",
    "rm runinfo*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Samples from Tanzania\n",
    "\n",
    "First subset of 8 samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bioproject\n",
    "bioproject=PRJNA439266\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir/$fln/\" ]] && mkdir -p \"$ldir/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: dowloading problem\" >> \"$ldir/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir/$fln/\" --split-files \"$ldir/$fln/$run\"\n",
    "    rm \"$ldir/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir/$fln/${run}_1.fastq\" \"$ldir/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir/$fln/${run}_2.fastq\" \"$ldir/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < <(tail -n +2 runinfo | sed \"/^$/d\")\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir/\"*/*\n",
    "\n",
    "rm runinfo*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Samples from Senegal, Niger and Tanzania"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bioproject\n",
    "bioproject=PRJNA560070\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir/$fln/\" ]] && mkdir -p \"$ldir/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: dowloading problem\" >> \"$ldir/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir/$fln/\" --split-files \"$ldir/$fln/$run\"\n",
    "    rm \"$ldir/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir/$fln/${run}_1.fastq\" \"$ldir/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir/$fln/${run}_2.fastq\" \"$ldir/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < <(tail -n +2 runinfo | sed \"/^$/d\")\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir/\"*/*\n",
    "\n",
    "rm runinfo*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Samples from Oman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bioproject\n",
    "bioproject=PRJNA560069\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir/$fln/\" ]] && mkdir -p \"$ldir/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: dowloading problem\" >> \"$ldir/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir/$fln/\" --split-files \"$ldir/$fln/$run\"\n",
    "    rm \"$ldir/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir/$fln/${run}_1.fastq\" \"$ldir/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir/$fln/${run}_2.fastq\" \"$ldir/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < <(tail -n +2 runinfo | sed \"/^$/d\")\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir/\"*/*\n",
    "\n",
    "rm runinfo*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Samples from Brazil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bioproject\n",
    "bioproject=PRJNA743359\n",
    "\n",
    "# Download related information to data project\n",
    "wget -q -O runinfo \"http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?save=efetch&rettype=runinfo&db=sra&term=${bioproject}\"\n",
    "\n",
    "# Field of interest (library name and weblink)\n",
    "fdn=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"SampleName\" | cut -d \":\" -f 1)\n",
    "fdr=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"Run\" | cut -d \":\" -f 1)\n",
    "flk=$(head -n 1 runinfo | tr \",\" \"\\n\" | grep -w -n \"download_path\" | cut -d \":\" -f 1)\n",
    "\n",
    "# Download fastq files\n",
    "while read line\n",
    "do\n",
    "    # Filename, run and download link\n",
    "    fln=$(cut -d \",\" -f $fdn <<<$line)\n",
    "    run=$(cut -d \",\" -f $fdr <<<$line)\n",
    "    lnk=$(cut -d \",\" -f $flk <<<$line)\n",
    "    \n",
    "    # Skip replicated libraries\n",
    "    [[ \"$fln\" =~ _rep$ ]] && continue\n",
    "    \n",
    "    # Download\n",
    "    echo \"$fln\"\n",
    "    [[ ! -d \"$ldir/$fln/\" ]] && mkdir -p \"$ldir/$fln/\"\n",
    "    retry=0\n",
    "    \n",
    "    while [[ $retry -lt 2 ]]\n",
    "    do\n",
    "        # Download sra file\n",
    "        wget -q -c -O \"$ldir/$fln/$run\" \"$lnk\"\n",
    "        # Check integrity\n",
    "        vdb-validate -q \"$ldir/$fln/$run\" &> /dev/null\n",
    "        [[ $? -ne 0 ]] && ((retry++)) || break\n",
    "    done\n",
    "    \n",
    "    # If max download attempt reached, issue message and move to the next\n",
    "    [[ $retry -eq 2 ]] && echo \"$run: downloading problem\" >> \"$ldir/download_issue\" && contine\n",
    "    \n",
    "    # Convert sra into fastq\n",
    "    fastq-dump -O \"$ldir/$fln/\" --split-files \"$ldir/$fln/$run\"\n",
    "    rm \"$ldir/$fln/$run\"\n",
    "    \n",
    "    # Rename file with more meaningful name\n",
    "    mv \"$ldir/$fln/${run}_1.fastq\" \"$ldir/$fln/${fln}_R1.fastq\"\n",
    "    mv \"$ldir/$fln/${run}_2.fastq\" \"$ldir/$fln/${fln}_R2.fastq\"\n",
    "        \n",
    "done < <(tail -n +2 runinfo | sed \"/^$/d\")\n",
    "\n",
    "# Compress files\n",
    "pigz \"$ldir/\"*/*\n",
    "\n",
    "rm runinfo*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genome data\n",
    "\n",
    "The genome data is downloaded from the [WormBase ParaSite](https://parasite.wormbase.org). We use the data from the version 14 (WBPS14) for *S. mansoni*. The data is then indexed by the different tools used in the downstream pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdir=\"data/genome\"\n",
    "[[ ! -d \"$gdir\" ]] && mkdir -p \"$gdir\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download and unzip data\n",
    "wget -P \"$gdir\" ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS14/species/schistosoma_mansoni/PRJEA36577/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa.gz\n",
    "pigz -d \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa.gz\"\n",
    "\n",
    "# Preparing indices\n",
    "bwa index \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "samtools faidx \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "gatk CreateSequenceDictionary -R \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "\n",
    "# Download the S. mansoni genome annotation\n",
    "wget -P \"$gdir\" ftp://ftp.ebi.ac.uk/pub/databases/wormbase/parasite/releases/WBPS14/species/schistosoma_mansoni/PRJEA36577/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3.gz\n",
    "pigz -d \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Known variants\n",
    "\n",
    "List of variants from the SmLE x SmHR crosses that segregated in a Mendelian fashion in F1 ([Valentim *et al.* 2013](https://doi.org/10.1126/science.1243106)). This list was generated from alignments using the v5 genome. This has been lifted over to the latest genome using the [flo pipeline](https://github.com/wurmlab/flo/tree/727f10b2b1c57a0514835d302d7f6345d3a34ffb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Index file\n",
    "gatk IndexFeatureFile -I \"$gdir/sm_dbSNP_v7.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For Freebayes\n",
    "bgzip -c \"$gdir/sm_dbSNP_v7.vcf\" > \"$gdir/sm_dbSNP_v7.vcf.gz\"\n",
    "tabix \"$gdir/sm_dbSNP_v7.vcf.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population files\n",
    "\n",
    "These populations files help to defined populations during variant calling with FreeBayes. Samples from Uganda were finally not included because of the absence of information regarding the gene of interest (too few reads mapped). However the line has only been commented in case one is interested in including this data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find \"$ldir\" -type d -name \"*SAM*\"   | sed \"s|.*/||g\" | sed \"s|$|\\tUganda|g\" > data/Sm_pop\n",
    "find \"$ldir\" -type d -name \"*Sm.TZ*\" | sed \"s|.*/||g\" | sed \"s|$|\\tTanzania|g\" >> data/Sm_pop\n",
    "find \"$ldir\" -type d -name \"*Sm.SN*\" | sed \"s|.*/||g\" | sed \"s|$|\\tSenegal|g\" >> data/Sm_pop\n",
    "find \"$ldir\" -type d -name \"*Sm.NE*\" | sed \"s|.*/||g\" | sed \"s|$|\\tNiger|g\" >> data/Sm_pop\n",
    "find \"$ldir\" -type d -name \"*Sm.BR*\" | sed \"s|.*/||g\" | sed \"s|$|\\tBrazil|g\" >> data/Sm_pop\n",
    "find \"$ldir\" -type d -name \"*Sm.OM*\" | sed \"s|.*/||g\" | sed \"s|$|\\tOman|g\" >> data/Sm_pop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene GFF files\n",
    "\n",
    "Gene specific GFF file is required for some downstream analyses. This generates a GFF for *Sm.TRPM$_{PZQ}$*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene code\n",
    "mygene=Smp_246790\n",
    "grep \"$mygene\" \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3\" > \"$gdir/$mygene.gff\"\n",
    "\n",
    "# Generate isoform specific gff\n",
    "myiso=$(grep -o \"$mygene\\.[0-9]*\" \"$gdir/$mygene.gff\" | sort | uniq)\n",
    "\n",
    "for i in $myiso\n",
    "do\n",
    "    awk '$3 == \"gene\"' \"$gdir/$mygene.gff\" > \"$gdir/$i.gff\"\n",
    "    grep -w \"$i\" \"$gdir/$mygene.gff\" | grep -v \"exon\" >> \"$gdir/$i.gff\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exome sequence data processing\n",
    "\n",
    "### Snakemake pipeline\n",
    "\n",
    "The snakemake pipeline performs the following steps:\n",
    "* alignment of each library against the reference genome\n",
    "* marking duplicates\n",
    "* base quality recalibration\n",
    "* generating alignment statistics\n",
    "* outputing Z chromosome read depth for each library\n",
    "* determining sex of each individual\n",
    "* outputing TRPM$_{PZQ}$ gene read depth for each library (*S. mansoni*: Smp_246790)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Status directory\n",
    "statdir=status\n",
    "[[ ! -d \"$statdir\" ]] && mkdir \"$statdir\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snakemake --snakefile snakefile --cluster \"qsub -V -cwd -o $statdir -j y -r y -pe smp 10 -S /bin/bash\" --jobs 1000 -w 120"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variant calling\n",
    "\n",
    "We call the variants within the gene region across all samples directly from the BAM files using FreeBayes. This method is preferred because it can perform population defined calling."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output directory\n",
    "cdir=\"data/variant_calling\"\n",
    "[[ ! -d \"$cdir\" ]] && mkdir -p \"$cdir\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Lenient calling\n",
    "\n",
    "We perform a lenient calling to identify all possible variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variables\n",
    "mygenome=\"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "#mypos=SM_V7_3:600000-1000000\n",
    "mypos=SM_V7_3:630000-800000\n",
    "\n",
    "# Switch the commented lines if samples from Uganda needs to be included\n",
    "# myfiles=$(find \"$ldir\" -name *SAM*.bam -o -name *Sm*.bam | sort | tr \"\\n\" \" \")\n",
    "myfiles=$(find \"$ldir\" -name *Sm*.bam | sort | tr \"\\n\" \" \")\n",
    "\n",
    "myfilename=\"$cdir/PZQ-R_field_lenient\"\n",
    "\n",
    "# Command used for calling variants\n",
    "freebayes -f \"$mygenome\" \\\n",
    "   -r \"$mypos\" \\\n",
    "   -b $myfiles \\\n",
    "   -=          \\\n",
    "   --population data/Sm_pop \\\n",
    "   -q 20       \\\n",
    "   -m 30       \\\n",
    "   -@ \"$gdir/sm_dbSNP_v7.vcf.gz\" > \"${myfilename}.vcf\"\n",
    "\n",
    "pigz \"${myfilename}\".vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Stringent calling\n",
    "\n",
    "We perform this second calling to identify variant likely to be real."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variables\n",
    "mygenome=\"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\"\n",
    "#mypos=SM_V7_3:600000-1000000\n",
    "mypos=SM_V7_3:630000-800000\n",
    "\n",
    "# Switch the commented lines if samples from Uganda needs to be included\n",
    "# myfiles=$(find \"$ldir\" -name *SAM*.bam -o -name *Sm*.bam | sort | tr \"\\n\" \" \")\n",
    "myfiles=$(find \"$ldir\" -name *Sm*.bam | sort | tr \"\\n\" \" \")\n",
    "\n",
    "myfilename=\"$cdir/PZQ-R_field_stringent\"\n",
    "\n",
    "# Command used for calling variants\n",
    "freebayes -f \"$mygenome\" \\\n",
    "   -r \"$mypos\" \\\n",
    "   -b $myfiles \\\n",
    "   -=          \\\n",
    "   --population data/Sm_pop \\\n",
    "   -q 20       \\\n",
    "   -m 30       \\\n",
    "   -F 0.3      \\\n",
    "   --min-coverage 10 \\\n",
    "   -@ \"$gdir/sm_dbSNP_v7.vcf.gz\" > \"${myfilename}.vcf\"\n",
    "\n",
    "pigz \"${myfilename}\".vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyzing SmTRPM$_{PZQ}$\n",
    "\n",
    "We analyze the previously called variants on the gene only. This analysis includes:\n",
    "* Filtering out genotypes with low read depth\n",
    "* Removing invariable sites\n",
    "* Normalizing coordinates\n",
    "* Use of a custom script to summarize variants and genotype from the VCF file\n",
    "\n",
    "\n",
    "### Variant calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir=\"results/1-reports\"\n",
    "[[ ! -d \"$rdir\" ]] && mkdir -p \"$rdir\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Lenient calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myfilename=\"$cdir/PZQ-R_field_lenient\"\n",
    "\n",
    "# Remove genotypes with low read depths\n",
    "vcftools --gzvcf \"${myfilename}.vcf.gz\" \\\n",
    "    --minDP 4 \\\n",
    "    --recode  \\\n",
    "    --recode-INFO-all \\\n",
    "    --out \"${myfilename}.flt-dp4\"\n",
    "\n",
    "# Rename file\n",
    "mv \"${myfilename}.flt-dp4.recode.vcf\" \"${myfilename}.flt-dp4.vcf\"\n",
    "\n",
    "# Clean the VCF of invariable ref variants\n",
    "vcf-subset -a -e \"${myfilename}.flt-dp4.vcf\" > \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "\n",
    "# Normalize VCF\n",
    "bcftools norm -o \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "    -f \"$mygenome\" \\\n",
    "    \"${myfilename}.flt-dp4.cleaned.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene and isoforms\n",
    "mygene=Smp_246790\n",
    "myiso=$(grep -o \"$mygene\\.[0-9]*\" \"$gdir/$mygene.gff\" | sort | uniq)\n",
    "\n",
    "# Generate reports per isoform\n",
    "for i in $myiso\n",
    "do\n",
    "    qsub -V -cwd -o \"$statdir\" -j y -b y scripts/func-table-mut.sh -v \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "        -r \"$mygenome\" \\\n",
    "        -g \"$gdir/$i.gff\" \\\n",
    "        -p data/Sm_pop \\\n",
    "        -o \"$rdir/$i.flt.norm.tsv\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Stringent calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myfilename=\"$cdir/PZQ-R_field_stringent\"\n",
    "\n",
    "# Remove genotypes with low read depths\n",
    "vcftools --gzvcf \"${myfilename}.vcf.gz\" \\\n",
    "    --minDP 4 \\\n",
    "    --recode  \\\n",
    "    --recode-INFO-all \\\n",
    "    --out \"${myfilename}.flt-dp4\"\n",
    "\n",
    "# Rename file\n",
    "mv \"${myfilename}.flt-dp4.recode.vcf\" \"${myfilename}.flt-dp4.vcf\"\n",
    "\n",
    "# Clean the VCF of invariable ref variants\n",
    "vcf-subset -a -e \"${myfilename}.flt-dp4.vcf\" > \"${myfilename}.flt-dp4.cleaned.vcf\"\n",
    "\n",
    "# Normalize VCF\n",
    "bcftools norm -o \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "    -f \"$mygenome\" \\\n",
    "    \"${myfilename}.flt-dp4.cleaned.vcf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene and isoforms\n",
    "mygene=Smp_246790\n",
    "myiso=$(grep -o \"$mygene\\.[0-9]*\" \"$gdir/$mygene.gff\" | sort | uniq)\n",
    "\n",
    "\n",
    "qsub -V -cwd -o \"$statdir\" -j y -b y scripts/func-table-mut.sh -v \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "        -r \"$mygenome\" \\\n",
    "        -g \"$gdir/$mygene.gff\" \\\n",
    "        -p data/Sm_pop \\\n",
    "        -o \"$rdir/${myfilename##*/}_${mygene}.flt.norm.tsv\"\n",
    "\n",
    "# Generate reports per isoform\n",
    "for i in $myiso\n",
    "do\n",
    "    qsub -V -cwd -o \"$statdir\" -j y -b y scripts/func-table-mut.sh -v \"${myfilename}.flt-dp4.cleaned.norm.vcf\" \\\n",
    "        -r \"$mygenome\" \\\n",
    "        -g \"$gdir/$i.gff\" \\\n",
    "        -p data/Sm_pop \\\n",
    "        -o \"$rdir/${myfilename##*/}_${i}.flt.norm.tsv\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Global coverage of the gene\n",
    "\n",
    "We inspect how well the gene and particularly the exons are covered. This is done by concatenating individual coverage of the gene and then plotting total read depth at each position. The concatenation step can take ~20 minutes or more."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir2=\"results/2-coverage\"\n",
    "[[ ! -d \"$rdir2\" ]] && mkdir -p \"$rdir2\"\n",
    "\n",
    "# Output file\n",
    "fl_out=\"$rdir2/SmTRP-PZQ.cov\"\n",
    "\n",
    "# List files\n",
    "fl_ls=($(find \"$ldir\" -type f -name \"*PZQ.cov\" | sort))\n",
    "\n",
    "# Create header\n",
    "echo -e \"#CHROM\\t$(cut -f1 \"${fl_ls[0]}\" | tr \"\\n\" \"\\t\" | sed \"s/\\t$//\")\" > \"$fl_out\"\n",
    "echo -e \"POS\\t$(cut -f2 \"${fl_ls[0]}\"| tr \"\\n\" \"\\t\" | sed \"s/\\t$//\")\" >> \"$fl_out\"\n",
    "\n",
    "for ((i = 0 ; i < ${#fl_ls[@]} ; i++))\n",
    "do\n",
    "    echo $i\n",
    "    echo -e \"$(dirname \"${fl_ls[$i]}\" | sed \"s,.*/,,\")\\t$(cut -f 3 \"${fl_ls[$i]}\" | tr \"\\n\" \"\\t\" | sed \"s/\\t$//\")\"  >> \"$fl_out\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting data\n",
    "Rscript scripts/SmTRP-PZQ_coverage.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutations of interest\n",
    "\n",
    "Some of the mutations are of high interest because located in critical sections of the protein (PZQ binding pocket, channel, TRP box). We list the samples that carry them in order to look at the reads within these samples to assess the likelihood of these mutations. This will also help identify samples that will require further investigation like PCR validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir3=\"results/3-mutations of interest\"\n",
    "[[ ! -d \"$rdir3\" ]] && mkdir -p \"$rdir3\"\n",
    "\n",
    "# Isoform of interest\n",
    "myiso=Smp_246790.5\n",
    "\n",
    "# mutations=\"p.T1394I p.Q1432H p.S1448Y p.G1458C p.H1522Q p.I1523N p.P1527T p.P1527Q p.Q1598K p.D1602Y p.D1606Y p.Q1609L p.Q1673K p.M1674I p.D1677Y p.H1680N p.P1683Q p.P1683R p.L1684I p.P1686T p.P1686Q p.P1687Q p.W1692L p.E1696* p.A1700D p.Q1704K\"\n",
    "mutations=\"p.H1522Q p.I1523N p.P1527T p.P1527Q p.Q1598K p.D1602Y p.D1606Y p.Q1609L p.Q1673K p.M1674I p.D1677Y p.H1680N p.P1686T p.P1686Q p.P1687Q\"\n",
    "\n",
    "# Add stop codon\n",
    "mutations+=\" $(awk '$8 ~ /*/ {print $8}' $rdir/${myiso}.flt.norm.tsv)\"\n",
    "\n",
    "unset mylist\n",
    "for m in $mutations\n",
    "do\n",
    "echo $m\n",
    "    mypos=$(awk -v m=$m '$8 == m {print $1}' \"$rdir/$myiso.flt.norm.tsv\")\n",
    "    myspl=$(awk -v mypos=$mypos '$2 == mypos' \"${myfilename}.flt-dp4.cleaned.norm.vcf\"| tr \"\\t\" \"\\n\" | grep -n 0/1 | cut -d \":\" -f 1 | tr \"\\n\" \",\" | sed \"s/,$//\")\n",
    "    myspl=$(grep -m 1 \"#C\" \"${myfilename}.flt-dp4.cleaned.norm.vcf\" | cut -f $myspl)\n",
    "    mylist+=\"$m\\t$mypos\\t$myspl\"\"\\n\" \n",
    "done\n",
    "\n",
    "echo -ne \"$mylist\" > \"$rdir3/samples_list.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output read depth for each sample carrying mutations of interest\n",
    "Rscript scripts/variant_read_depth.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plots of the read depth for each mutation and sample are available in the `graphs` folder."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequences for primer design\n",
    "\n",
    "Interesting mutations from the calling data must be checked by Sanger sequencing. These mutations are located in specific exons. To sequence the full exons, we extract exon sequences with an additional 200 bp on each end, where primers should be designed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fl_out=\"$rdir3/exons_of_interest\"\n",
    "\n",
    "# Exons number refers to the isoform 5 and not the entire gene\n",
    "myexons=(3 4 12 23 25 27 29 34)\n",
    "myiso=\"Smp_246790.5\"\n",
    "\n",
    "grep \"$myiso\" \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3\" | \\\n",
    "    awk '$3 == \"CDS\"'  | \\\n",
    "    sed -n \"$(sed \"s/ /p;/g\" <<< ${myexons[@]})p\" | \\\n",
    "    awk '{print $1\"\\t\"$4-201\"\\t\"$5+200}' > \"$fl_out.bed\"\n",
    "\n",
    "bedtools getfasta -fi \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\" \\\n",
    "    -bed \"$fl_out.bed\" \\\n",
    "    -fo \"$fl_out.fas\"\n",
    "\n",
    "# Reformat fasta entry\n",
    "myentries=($(grep -n \">\" \"$fl_out.fas\" | cut -d \":\" -f 1))\n",
    "for ((i = 0 ; i < \"${#myexons[@]}\" ; i++))\n",
    "do\n",
    "    sed -i \"${myentries[$i]}s/>/>E${myexons[$i]} /\" \"$fl_out.fas\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Primers are then designed using PerlPrimer. See the publication for the list of primers.\n",
    "\n",
    "\n",
    "### Sanger sequencing\n",
    "\n",
    "Sanger sequencing traces are analyzed using Consed and PolyPhred: the first align the data while the second perform the calling. PolyPhred output is then processed with a custom script in order to summarize the calling per sample.\n",
    "\n",
    "**Note**: Sanger sequencing traces have been deposited on [Zenodo](https://doi.org/10.5281/zenodo.5204523)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data directory\n",
    "sdir=\"data/sanger_sequencing\"\n",
    "[[ ! -d \"$sdir\" ]] && mkdir -p \"$sdir\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Download the data from Zenodo\n",
    "wget -P \"$sdir\" \"https://zenodo.org/record/5204523/files/traces.tar.gz\"\n",
    "\n",
    "# Extract files\n",
    "tar xzvf \"$sdir/traces.tar.gz\" -C \"$sdir\"\n",
    "\n",
    "# Remove archive\n",
    "rm \"$sdir/traces.tar.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Result folder\n",
    "rdir3b=\"$rdir3/sanger_sequencing\"\n",
    "[[ ! -d \"$rdir3b\" ]] && mkdir -p \"$rdir3b\"\n",
    "\n",
    "# Phred folders\n",
    "mkdir -p \"$rdir3b/data/\"{chromat_dir,edit_dir,phd_dir,poly_dir}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mygene=\"Smp_246790\"\n",
    "\n",
    "# Generate reference sequence\n",
    "# mkdir \"$rdir3b/0-Ref_seq\"\n",
    "bedtools getfasta -fi \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.genomic.fa\" \\\n",
    "    -bed <(grep \"$mygene\" \"$gdir/schistosoma_mansoni.PRJEA36577.WBPS14.annotations.gff3\" | awk '$3 == \"gene\"' | awk '{print $1\"\\t\"$4-1\"\\t\"$5}') \\\n",
    "    -fo \"$rdir3b/Smp_246790.fas\"\n",
    "\n",
    "# Rename fasta entry\n",
    "sed -i \"s/>.*/>$mygene/\" \"$rdir3b/Smp_246790.fas\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pwd_old=$PWD\n",
    "\n",
    "# Generate ref sequence for phred\n",
    "cd \"$rdir3b/\"\n",
    "sudophred \"$mygene.fas\" -r -abi\n",
    "\n",
    "# Move ref\n",
    "mv \"$mygene.REF\" \"data/chromat_dir/\"\n",
    "mv \"$mygene.REF.phd.1\" \"data/phd_dir/\"\n",
    "\n",
    "cd \"$pwd_old\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Preparing data\n",
    "\n",
    "We create links to the original data but with modified file names in order to inform Polyphred about the paired samples. Samples will be identified on the first 8 characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Link barcode list\n",
    "ln -s \"$PWD/$sdir/barcode_list.tsv\" \"$rdir3b\"\n",
    "\n",
    "# Generate modified sample names for polyphred\n",
    "j=1\n",
    "unset spl_code\n",
    "for i in $(tail -n +2 \"$rdir3b/barcode_list.tsv\" | cut -f 1 | sort | uniq)\n",
    "do\n",
    "    spl_code+=\"$i\\t$(sed \"s/^/$(printf \"%02d\" $j)_/\" <<< \"$i\")\\n\"\n",
    "    ((j++))\n",
    "done\n",
    "\n",
    "# Generate exon specific folder and link to data\n",
    "while read line \n",
    "do\n",
    "    # Sample name\n",
    "    myspl=$(cut -f 1 <<< \"$line\")\n",
    "    myspl=$(echo -ne \"$spl_code\" | grep -w \"$myspl\" | cut -f2)\n",
    "    \n",
    "    # Barcode\n",
    "    mycode=$(cut -f 3 <<< \"$line\")\n",
    "    [[ \"$mycode\" == NA ]] && continue\n",
    "    \n",
    "    # Link files\n",
    "    ln -s \"$PWD/$sdir/traces/$mycode\"*.scf \"$rdir3b/data/chromat_dir/${myspl}_$mycode.scf\"\n",
    "\n",
    "done < <(tail -n +2 \"$rdir3b/barcode_list.tsv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Running Phred and Consed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pwd_old=$PWD\n",
    "\n",
    "# Moving to the working folder\n",
    "cd \"$rdir3b/data/edit_dir\"\n",
    "\n",
    "# Generating phd and poly files\n",
    "phred ../chromat_dir/*.scf -pd ../phd_dir/ -dd ../poly_dir/\n",
    "\n",
    "# Preparing the reference sequence\n",
    "phd2Ace.perl \"$mygene.REF.phd.1\"\n",
    "\n",
    "# Aligning reads to the reference\n",
    "find ../chromat_dir/* -name \"*.scf\" -exec basename {} \\; > reads_to_add.fof\n",
    "consed -ace \"$mygene.REF.ace\" -addNewReads reads_to_add.fof -newAceFilename \"$mygene.REF.ace.1\"\n",
    "cd \"$pwd_old\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Running PolyPhred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "score_snp=70 #90\n",
    "score_indel=70 #80\n",
    "qual=25 #40\n",
    "\n",
    "# SNP scoring\n",
    "name=$mygene.s${score_snp}-q$qual\n",
    "polyphred -d \"$rdir3b/data\" -extended_genotype -n $name.snp.nav -o $name.snp.polyphred.out -q $qual -refcomp -s 1 8 -score $score_snp\n",
    "\n",
    "# Indel scoring\n",
    "name=$mygene.s${score_indel}-q$qual\n",
    "polyphred -d \"$rdir3b/data\" -extended_genotype -i -inav $name.indel.nav -o $name.indel.polyphred.out -q $qual -s 1 8 -iscore $score_indel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To summarize the finding from PolyPhred, we generate a table using the custom script below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genotype extractions\n",
    "scripts/trim_polyphred.sh -i \"$rdir3b/data/edit_dir/$mygene.s${score_snp}-q$qual.snp.polyphred.out\" \\\n",
    "    -o \"$rdir3b/$mygene.SNP.gt.tsv\" \\\n",
    "    -p 2 3 \\\n",
    "    -d \"_\"\n",
    "\n",
    "scripts/trim_polyphred.sh -i \"$rdir3b/data/edit_dir/$mygene.s${score_indel}-q$qual.indel.polyphred.out\" \\\n",
    "    -o \"$rdir3b/$mygene.INDEL.gt.tsv\" \\\n",
    "    -p 2 3 \\\n",
    "    -d \"_\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results show that non of the tested mutations have been confirmed through Sanger sequencing."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "384px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
